// @aztec/barretenberg/dest/fft/single_fft.js
import { Transfer } from '../transport/index.js';
export class SingleFft {
    constructor(wasm) {
        this.wasm = wasm;
    }
    async init(circuitSize) {
        this.domainPtr = await this.wasm.call('new_evaluation_domain', circuitSize);
    }
    async destroy() {
        await this.wasm.call('delete_evaluation_domain', this.domainPtr);
    }
    async fft(coefficients, constant) {
        const circuitSize = coefficients.length / 32;
        const newPtr = await this.wasm.call('bbmalloc', coefficients.length);
        await this.wasm.transferToHeap(Transfer(coefficients, [coefficients.buffer]), newPtr);
        await this.wasm.transferToHeap(Transfer(constant, [constant.buffer]), 0);
        await this.wasm.call('coset_fft_with_generator_shift', newPtr, 0, this.domainPtr);
        const result = await this.wasm.sliceMemory(newPtr, newPtr + circuitSize * 32);
        await this.wasm.call('bbfree', newPtr);
        return result;
    }
    async ifft(coefficients) {
        const circuitSize = coefficients.length / 32;
        const newPtr = await this.wasm.call('bbmalloc', coefficients.length);
        await this.wasm.transferToHeap(Transfer(coefficients, [coefficients.buffer]), newPtr);
        await this.wasm.call('ifft', newPtr, this.domainPtr);
        const result = await this.wasm.sliceMemory(newPtr, newPtr + circuitSize * 32);
        await this.wasm.call('bbfree', newPtr);
        return result;
    }
}
export class SingleFftFactory {
    constructor(wasm) {
        this.wasm = wasm;
        this.ffts = {};
    }
    async createFft(circuitSize) {
        if (!this.ffts[circuitSize]) {
            const fft = new SingleFft(this.wasm);
            await fft.init(circuitSize);
            this.ffts[circuitSize] = fft;
        }
        return this.ffts[circuitSize];
    }
    async destroy() { }
}
