// @aztec/sdk/dest/aztec_sdk/aztec_sdk_user.js
export class AztecSdkUser {
    constructor(id, sdk) {
        this.id = id;
        this.sdk = sdk;
    }
    async isSynching() {
        return await this.sdk.isUserSynching(this.id);
    }
    async awaitSynchronised(timeout) {
        return await this.sdk.awaitUserSynchronised(this.id, timeout);
    }
    async getSyncedToRollup() {
        return await this.sdk.getUserSyncedToRollup(this.id);
    }
    async getSpendingKeys() {
        return await this.sdk.getSpendingKeys(this.id);
    }
    async getBalances() {
        return await this.sdk.getBalances(this.id);
    }
    async getBalance(assetId) {
        return await this.sdk.getBalance(this.id, assetId);
    }
    async getSpendableSum(assetId, spendingKeyRequired, excludePendingNotes) {
        return await this.sdk.getSpendableSum(this.id, assetId, spendingKeyRequired, excludePendingNotes);
    }
    async getSpendableSums(spendingKeyRequired, excludePendingNotes) {
        return await this.sdk.getSpendableSums(this.id, spendingKeyRequired, excludePendingNotes);
    }
    async getMaxSpendableValue(assetId, spendingKeyRequired, excludePendingNotes, numNotes) {
        return await this.sdk.getMaxSpendableValue(this.id, assetId, spendingKeyRequired, excludePendingNotes, numNotes);
    }
    async getTxs() {
        return await this.sdk.getUserTxs(this.id);
    }
    async getPaymentTxs() {
        return await this.sdk.getPaymentTxs(this.id);
    }
    async getAccountTxs() {
        return await this.sdk.getAccountTxs(this.id);
    }
    async getDefiTxs() {
        return await this.sdk.getDefiTxs(this.id);
    }
}
