// @aztec/sdk/dest/aztec_sdk/create_aztec_sdk.js
import { createDebugLogger, enableLogs, logHistory } from '@aztec/barretenberg/log';
import { ClientEthereumBlockchain } from '@aztec/blockchain';
import { createCoreSdk } from '../core_sdk/index.js';
import { AztecSdk } from './aztec_sdk.js';
const debug = createDebugLogger('bb:create_aztec_sdk');
async function createBlockchain(ethereumProvider, coreSdk, confs = 3) {
    const { chainId, rollupContractAddress, permitHelperContractAddress } = await coreSdk.getLocalStatus();
    const { blockchainStatus: { assets, bridges }, } = await coreSdk.getRemoteStatus();
    const blockchain = new ClientEthereumBlockchain(rollupContractAddress, permitHelperContractAddress, assets, bridges, ethereumProvider, confs);
    const providerChainId = await blockchain.getChainId();
    if (chainId !== providerChainId) {
        throw new Error(`Provider chainId ${providerChainId} does not match rollup provider chainId ${chainId}.`);
    }
    return blockchain;
}
// TODO - remove it
export var SdkFlavour;
(function (SdkFlavour) {
    SdkFlavour[SdkFlavour["PLAIN"] = 0] = "PLAIN";
    SdkFlavour[SdkFlavour["SHARED_WORKER"] = 1] = "SHARED_WORKER";
    SdkFlavour[SdkFlavour["HOSTED"] = 2] = "HOSTED";
})(SdkFlavour = SdkFlavour || (SdkFlavour = {}));
export async function createAztecSdk(ethereumProvider, options) {
    if (options.flavour) {
        console.warn('SdkFlavour has been deprecated.');
    }
    if (options.debug) {
        enableLogs(options.debug);
        logHistory.enable();
    }
    const coreSdk = await createCoreSdk(options);
    try {
        const blockchain = await createBlockchain(ethereumProvider, coreSdk, options.minConfirmation);
        return new AztecSdk(coreSdk, blockchain, ethereumProvider);
    }
    catch (err) {
        debug(`failed to create sdk: ${err.message}`);
        await coreSdk.destroy();
        throw err;
    }
}
