// @aztec/blockchain/dest/contracts/rollup_processor/proxy_admin.js
import { EthAddress } from '@aztec/barretenberg/address';
import { Contract, ContractFactory } from 'ethers';
import { formatBytes32String } from 'ethers/lib/utils.js';
import { ProxyAdminContract, ProxyContract, ProxyDeployer } from '../../abis.js';
/**
 * Thin wrapper around the Proxy Admin contract.
 */
export class ProxyAdmin {
    constructor(signer) {
        this.signer = signer;
        this.vanitySalt = formatBytes32String('Aztec Connect');
        this.proxyAdmin = new Contract(EthAddress.ZERO.toString(), ProxyAdminContract.abi, this.signer);
        this.proxyDeployer = new Contract(EthAddress.ZERO.toString(), ProxyDeployer.abi, this.signer);
    }
    connectNewSigner(signer) {
        this.signer = signer;
        this.connectToInstance(this.address);
    }
    connectToInstance(address) {
        this.proxyAdmin = new Contract(address.toString(), ProxyAdminContract.abi, this.signer);
    }
    async deployInstance() {
        const factory = new ContractFactory(ProxyAdminContract.abi, ProxyAdminContract.bytecode, this.signer);
        this.proxyAdmin = await (await factory.deploy()).deployed();
        const deployerFactory = new ContractFactory(ProxyDeployer.abi, ProxyDeployer.bytecode, this.signer);
        this.proxyDeployer = await (await deployerFactory.deploy()).deployed();
    }
    get address() {
        return EthAddress.fromString(this.proxyAdmin.address);
    }
    get contract() {
        return this.proxyAdmin;
    }
    async owner() {
        return EthAddress.fromString(await this.proxyAdmin.owner());
    }
    async getProxyAdmin(proxy) {
        return await this.proxyAdmin.getProxyAdmin(proxy.toString());
    }
    async getProxyImplementation(proxy) {
        return await this.proxyAdmin.getProxyImplementation(proxy.toString());
    }
    async changeProxyAdmin(proxy, newAdmin) {
        await this.proxyAdmin.changeProxyAdmin(proxy.toString(), newAdmin.toString());
    }
    async transferProxyAdminOwnership(newOwner) {
        await this.proxyAdmin.transferOwnership(newOwner.toString());
    }
    async upgradeUNSAFE(proxy, implementationFactory, constructorArgs) {
        const implementation = await (await implementationFactory.deploy(...constructorArgs)).deployed();
        await this.proxyAdmin.upgrade(proxy.toString(), implementation.address);
        return implementation.attach(proxy.toString());
    }
    async upgradeAndInitializeWithConstructor(proxy, implementationFactory, initializeArgs, constructorArgs) {
        const implementation = await (await implementationFactory.deploy(...constructorArgs)).deployed();
        const calldata = implementation.interface.encodeFunctionData('initialize', initializeArgs);
        await this.proxyAdmin.upgradeAndCall(proxy.toString(), implementation.address, calldata);
        return implementation.attach(proxy.toString());
    }
    async deployProxyUNSAFE(implementationFactory, constructorArgs) {
        const implementation = await (await implementationFactory.deploy(...constructorArgs)).deployed();
        const proxyFactory = new ContractFactory(ProxyContract.abi, ProxyContract.bytecode, this.signer);
        const calldata = '0x';
        const proxy = await (await proxyFactory.deploy(implementation.address, this.proxyAdmin.address, calldata)).deployed();
        return implementation.attach(proxy.address);
    }
    async deployProxyAndInitializeWithConstructor(implementationFactory, initializeArgs, constructorArgs) {
        const implementation = await (await implementationFactory.deploy(...constructorArgs)).deployed();
        const calldata = implementation.interface.encodeFunctionData('initialize', initializeArgs);
        const tx = await (await this.proxyDeployer.deployProxy(implementation.address, this.proxyAdmin.address, calldata, this.vanitySalt)).wait();
        const events = tx.events;
        const proxyDeployedEvent = events[events.length - 1];
        const proxyAddress = proxyDeployedEvent.args['proxy'];
        const proxy = new Contract(proxyAddress, ProxyContract.abi, this.signer);
        return implementation.attach(proxy.address);
    }
}
