// @aztec/blockchain/dest/client_ethereum_blockchain.js
import { EthAddress } from '@aztec/barretenberg/address';
import { sleep } from '@aztec/barretenberg/sleep';
import { Timer } from '@aztec/barretenberg/timer';
import { Web3Provider } from '@ethersproject/providers';
import { EthAsset, RollupProcessor, TokenAsset } from './contracts/index.js';
export class ClientEthereumBlockchain {
    constructor(rollupContractAddress, permitHelperContractAddress, assets, bridges, ethereumProvider, minConfirmations, permitSupportAssetIds = []) {
        this.bridges = bridges;
        this.ethereumProvider = ethereumProvider;
        this.minConfirmations = minConfirmations;
        this.permitSupportAssetIds = permitSupportAssetIds;
        this.rollupProcessor = new RollupProcessor(rollupContractAddress, ethereumProvider, permitHelperContractAddress);
        this.provider = new Web3Provider(this.ethereumProvider);
        this.assets = assets.map(asset => {
            if (asset.address.equals(EthAddress.ZERO)) {
                return new EthAsset(this.ethereumProvider);
            }
            else {
                return TokenAsset.new(asset, this.ethereumProvider);
            }
        });
    }
    async getChainId() {
        return (await this.provider.getNetwork()).chainId;
    }
    getAsset(assetId) {
        return this.assets[assetId];
    }
    getAssetIdByAddress(address, gasLimit) {
        const assetId = this.assets.findIndex(a => a.getStaticInfo().address.equals(address) &&
            (gasLimit === undefined || a.getStaticInfo().gasLimit === gasLimit));
        if (assetId < 0) {
            throw new Error(`Unknown asset. Address: ${address}. Gas limit: ${gasLimit}.`);
        }
        return assetId;
    }
    getAssetIdBySymbol(symbol, gasLimit) {
        const assetId = this.assets.findIndex(a => a.getStaticInfo().symbol.toLowerCase() === symbol.toLowerCase() &&
            (gasLimit === undefined || a.getStaticInfo().gasLimit === gasLimit));
        if (assetId < 0) {
            throw new Error(`Unknown asset. Symbol: ${symbol}. Gas limit: ${gasLimit}.`);
        }
        return assetId;
    }
    getBridgeAddressId(address, gasLimit) {
        const index = this.bridges.findIndex(b => b.address.equals(address) && (gasLimit === undefined || b.gasLimit === gasLimit));
        if (index < 0) {
            throw new Error(`Unknown bridge. Address: ${address}. Gas limit: ${gasLimit}.`);
        }
        return index + 1;
    }
    hasPermitSupport(assetId) {
        return this.permitSupportAssetIds.includes(assetId);
    }
    async getUserPendingDeposit(assetId, account) {
        return await this.rollupProcessor.getUserPendingDeposit(assetId, account);
    }
    async getUserProofApprovalStatus(account, txId) {
        return await this.rollupProcessor.getProofApprovalStatus(account, txId);
    }
    async depositPendingFunds(assetId, amount, proofHash, options) {
        return await this.rollupProcessor.depositPendingFunds(assetId, amount, proofHash, options);
    }
    async depositPendingFundsPermit(assetId, amount, deadline, signature, options) {
        return await this.rollupProcessor.depositPendingFundsPermit(assetId, amount, deadline, signature, options);
    }
    async depositPendingFundsPermitNonStandard(assetId, amount, nonce, deadline, signature, options) {
        return await this.rollupProcessor.depositPendingFundsPermitNonStandard(assetId, amount, nonce, deadline, signature, options);
    }
    async approveProof(txId, options) {
        return await this.rollupProcessor.approveProof(txId, options);
    }
    async processAsyncDefiInteraction(interactionNonce, options) {
        return await this.rollupProcessor.processAsyncDefiInteraction(interactionNonce, options);
    }
    async isContract(address) {
        return (await this.provider.getCode(address.toString())) !== '0x';
    }
    async isEmpty(address) {
        return (!(await this.isContract(address)) &&
            (await this.provider.getBalance(address.toString())).toBigInt() == BigInt(0) &&
            (await this.provider.getTransactionCount(address.toString())) == 0);
    }
    async setSupportedAsset(assetAddress, assetGasLimit, options) {
        const txHash = await this.rollupProcessor.setSupportedAsset(assetAddress, assetGasLimit, options);
        return txHash;
    }
    async setSupportedBridge(bridgeAddress, bridgeGasLimit, options) {
        const txHash = await this.rollupProcessor.setSupportedBridge(bridgeAddress, bridgeGasLimit, options);
        return txHash;
    }
    /**
     * Wait for given transaction to be mined, and return receipt.
     */
    async getTransactionReceipt(txHash, timeout = 300, interval = 1, minConfirmations = this.minConfirmations) {
        const timer = new Timer();
        while (true) {
            const txReceipt = await this.provider.getTransactionReceipt(txHash.toString());
            if (!minConfirmations || (txReceipt && txReceipt.confirmations >= minConfirmations)) {
                return txReceipt
                    ? { status: !!txReceipt.status, blockNum: txReceipt.blockNumber }
                    : { status: false, blockNum: 0 };
            }
            await sleep(interval * 1000);
            if (timeout && timer.s() > timeout) {
                throw new Error(`Timeout awaiting tx confirmation: ${txHash}`);
            }
        }
    }
}
