// @aztec/blockchain/dest/create_permit_data.js
const types = {
    EIP712Domain: [
        { name: 'name', type: 'string' },
        { name: 'version', type: 'string' },
        { name: 'chainId', type: 'uint256' },
        { name: 'verifyingContract', type: 'address' },
    ],
    Permit: [
        {
            name: 'owner',
            type: 'address',
        },
        {
            name: 'spender',
            type: 'address',
        },
        {
            name: 'value',
            type: 'uint256',
        },
        {
            name: 'nonce',
            type: 'uint256',
        },
        {
            name: 'deadline',
            type: 'uint256',
        },
    ],
};
const noneStandardTypes = {
    EIP712Domain: [
        { name: 'name', type: 'string' },
        { name: 'version', type: 'string' },
        { name: 'chainId', type: 'uint256' },
        { name: 'verifyingContract', type: 'address' },
    ],
    Permit: [
        {
            name: 'holder',
            type: 'address',
        },
        {
            name: 'spender',
            type: 'address',
        },
        {
            name: 'nonce',
            type: 'uint256',
        },
        {
            name: 'expiry',
            type: 'uint256',
        },
        {
            name: 'allowed',
            type: 'bool',
        },
    ],
};
export function createPermitData(name, owner, spender, value, nonce, deadline, verifyingContract, chainId, version = '1') {
    const domain = {
        name,
        version,
        chainId: chainId.toString(),
        verifyingContract: verifyingContract.toString(),
    };
    const message = {
        owner: owner.toString(),
        spender: spender.toString(),
        value: value.toString(),
        nonce: nonce.toString(),
        deadline: deadline.toString(),
    };
    return { types, domain, message, primaryType: 'Permit' };
}
export function createPermitDataNonStandard(name, owner, spender, nonce, deadline, verifyingContract, chainId, version = '1') {
    const domain = {
        name,
        version,
        chainId: chainId.toString(),
        verifyingContract: verifyingContract.toString(),
    };
    const message = {
        holder: owner.toString(),
        spender: spender.toString(),
        nonce: nonce.toString(),
        expiry: deadline.toString(),
        allowed: true,
    };
    return { types: noneStandardTypes, domain, message, primaryType: 'Permit' };
}
