// @aztec/barretenberg/dest/offchain_tx_data/offchain_defi_deposit_data.js
import { GrumpkinAddress } from '../address/index.js';
import { toBigIntBE, toBufferBE } from '../bigint_buffer/index.js';
import { BridgeCallData } from '../bridge_call_data/index.js';
import { numToUInt32BE } from '../serialize/index.js';
import { ViewingKey } from '../viewing_key/index.js';
var DataSizes;
(function (DataSizes) {
    DataSizes[DataSizes["BRIDGE_CALL_DATA"] = BridgeCallData.SIZE] = "BRIDGE_CALL_DATA";
    DataSizes[DataSizes["PARTIAL_STATE"] = 32] = "PARTIAL_STATE";
    DataSizes[DataSizes["PARTIAL_STATE_SECRET_EPH_PUB_KEY"] = GrumpkinAddress.SIZE] = "PARTIAL_STATE_SECRET_EPH_PUB_KEY";
    DataSizes[DataSizes["DEPOSIT_VALUE"] = 32] = "DEPOSIT_VALUE";
    DataSizes[DataSizes["TX_FEE"] = 32] = "TX_FEE";
    DataSizes[DataSizes["VIEWING_KEY"] = ViewingKey.SIZE] = "VIEWING_KEY";
    DataSizes[DataSizes["TX_REF_NO"] = 4] = "TX_REF_NO";
})(DataSizes || (DataSizes = {}));
var DataOffsets;
(function (DataOffsets) {
    DataOffsets[DataOffsets["BRIDGE_CALL_DATA"] = 0] = "BRIDGE_CALL_DATA";
    DataOffsets[DataOffsets["PARTIAL_STATE"] = DataOffsets.BRIDGE_CALL_DATA + DataSizes.BRIDGE_CALL_DATA] = "PARTIAL_STATE";
    DataOffsets[DataOffsets["PARTIAL_STATE_SECRET_EPH_PUB_KEY"] = DataOffsets.PARTIAL_STATE + DataSizes.PARTIAL_STATE] = "PARTIAL_STATE_SECRET_EPH_PUB_KEY";
    DataOffsets[DataOffsets["DEPOSIT_VALUE"] = DataOffsets.PARTIAL_STATE_SECRET_EPH_PUB_KEY + DataSizes.PARTIAL_STATE_SECRET_EPH_PUB_KEY] = "DEPOSIT_VALUE";
    DataOffsets[DataOffsets["TX_FEE"] = DataOffsets.DEPOSIT_VALUE + DataSizes.DEPOSIT_VALUE] = "TX_FEE";
    DataOffsets[DataOffsets["VIEWING_KEY"] = DataOffsets.TX_FEE + DataSizes.TX_FEE] = "VIEWING_KEY";
    DataOffsets[DataOffsets["TX_REF_NO"] = DataOffsets.VIEWING_KEY + DataSizes.VIEWING_KEY] = "TX_REF_NO";
})(DataOffsets || (DataOffsets = {}));
export class OffchainDefiDepositData {
    constructor(bridgeCallData, partialState, partialStateSecretEphPubKey, // the public key from which the partial state's secret may be derived (when combined with a valid account private key).
    depositValue, txFee, viewingKey, // viewing key for the 'claim' note
    txRefNo = 0) {
        this.bridgeCallData = bridgeCallData;
        this.partialState = partialState;
        this.partialStateSecretEphPubKey = partialStateSecretEphPubKey;
        this.depositValue = depositValue;
        this.txFee = txFee;
        this.viewingKey = viewingKey;
        this.txRefNo = txRefNo;
        if (partialState.length !== DataSizes.PARTIAL_STATE) {
            throw new Error('Expect partialState to be 32 bytes.');
        }
        if (viewingKey.isEmpty()) {
            throw new Error('Viewing key cannot be empty.');
        }
    }
    static getViewingKeyBuffer(buf) {
        return buf.slice(DataOffsets.VIEWING_KEY, DataOffsets.VIEWING_KEY + DataSizes.VIEWING_KEY);
    }
    static fromBuffer(buf) {
        if (buf.length !== OffchainDefiDepositData.SIZE) {
            throw new Error('Invalid buffer size.');
        }
        const bridgeCallData = BridgeCallData.fromBuffer(buf.slice(DataOffsets.BRIDGE_CALL_DATA, DataOffsets.BRIDGE_CALL_DATA + DataSizes.BRIDGE_CALL_DATA));
        const partialState = buf.slice(DataOffsets.PARTIAL_STATE, DataOffsets.PARTIAL_STATE + DataSizes.PARTIAL_STATE);
        const partialStateSecretEphPubKey = new GrumpkinAddress(buf.slice(DataOffsets.PARTIAL_STATE_SECRET_EPH_PUB_KEY, DataOffsets.PARTIAL_STATE_SECRET_EPH_PUB_KEY + DataSizes.PARTIAL_STATE_SECRET_EPH_PUB_KEY));
        const depositValue = toBigIntBE(buf.slice(DataOffsets.DEPOSIT_VALUE, DataOffsets.DEPOSIT_VALUE + DataSizes.DEPOSIT_VALUE));
        const txFee = toBigIntBE(buf.slice(DataOffsets.TX_FEE, DataOffsets.TX_FEE + DataSizes.TX_FEE));
        const viewingKey = new ViewingKey(buf.slice(DataOffsets.VIEWING_KEY, DataOffsets.VIEWING_KEY + DataSizes.VIEWING_KEY));
        const txRefNo = buf.readUInt32BE(DataOffsets.TX_REF_NO);
        return new OffchainDefiDepositData(bridgeCallData, partialState, partialStateSecretEphPubKey, depositValue, txFee, viewingKey, txRefNo);
    }
    toBuffer() {
        return Buffer.concat([
            this.bridgeCallData.toBuffer(),
            this.partialState,
            this.partialStateSecretEphPubKey.toBuffer(),
            toBufferBE(this.depositValue, 32),
            toBufferBE(this.txFee, 32),
            this.viewingKey.toBuffer(),
            numToUInt32BE(this.txRefNo),
        ]);
    }
}
OffchainDefiDepositData.EMPTY = new OffchainDefiDepositData(BridgeCallData.ZERO, Buffer.alloc(32), GrumpkinAddress.ZERO, BigInt(0), BigInt(0), new ViewingKey(Buffer.alloc(ViewingKey.SIZE)));
OffchainDefiDepositData.SIZE = OffchainDefiDepositData.EMPTY.toBuffer().length;
