// @aztec/barretenberg/dest/offchain_tx_data/offchain_join_split_data.js
import { numToUInt32BE } from '../serialize/index.js';
import { ViewingKey } from '../viewing_key/index.js';
var DataSizes;
(function (DataSizes) {
    DataSizes[DataSizes["VIEWING_KEY_0"] = ViewingKey.SIZE] = "VIEWING_KEY_0";
    DataSizes[DataSizes["VIEWING_KEY_1"] = ViewingKey.SIZE] = "VIEWING_KEY_1";
    DataSizes[DataSizes["TX_REF_NO"] = 4] = "TX_REF_NO";
})(DataSizes || (DataSizes = {}));
var DataOffsets;
(function (DataOffsets) {
    DataOffsets[DataOffsets["VIEWING_KEY_0"] = 0] = "VIEWING_KEY_0";
    DataOffsets[DataOffsets["VIEWING_KEY_1"] = DataOffsets.VIEWING_KEY_0 + DataSizes.VIEWING_KEY_0] = "VIEWING_KEY_1";
    DataOffsets[DataOffsets["TX_REF_NO"] = DataOffsets.VIEWING_KEY_1 + DataSizes.VIEWING_KEY_1] = "TX_REF_NO";
})(DataOffsets || (DataOffsets = {}));
export class OffchainJoinSplitData {
    constructor(viewingKeys, txRefNo = 0) {
        this.viewingKeys = viewingKeys;
        this.txRefNo = txRefNo;
        if (viewingKeys.length !== 2) {
            throw new Error(`Expect 2 viewing keys. Received ${viewingKeys.length}.`);
        }
        if (viewingKeys.some(vk => vk.isEmpty())) {
            throw new Error('Viewing key cannot be empty.');
        }
    }
    static getViewingKeyBuffers(buf) {
        const viewingKey0 = buf.slice(DataOffsets.VIEWING_KEY_0, DataOffsets.VIEWING_KEY_0 + DataSizes.VIEWING_KEY_0);
        const viewingKey1 = buf.slice(DataOffsets.VIEWING_KEY_1, DataOffsets.VIEWING_KEY_1 + DataSizes.VIEWING_KEY_1);
        return [viewingKey0, viewingKey1];
    }
    static fromBuffer(buf) {
        if (buf.length !== OffchainJoinSplitData.SIZE) {
            throw new Error('Invalid buffer size.');
        }
        const viewingKey0 = new ViewingKey(buf.slice(DataOffsets.VIEWING_KEY_0, DataOffsets.VIEWING_KEY_0 + DataSizes.VIEWING_KEY_0));
        const viewingKey1 = new ViewingKey(buf.slice(DataOffsets.VIEWING_KEY_1, DataOffsets.VIEWING_KEY_1 + DataSizes.VIEWING_KEY_1));
        const txRefNo = buf.readUInt32BE(DataOffsets.TX_REF_NO);
        return new OffchainJoinSplitData([viewingKey0, viewingKey1], txRefNo);
    }
    toBuffer() {
        return Buffer.concat([...this.viewingKeys.map(k => k.toBuffer()), numToUInt32BE(this.txRefNo)]);
    }
}
OffchainJoinSplitData.EMPTY = new OffchainJoinSplitData([
    new ViewingKey(Buffer.alloc(ViewingKey.SIZE)),
    new ViewingKey(Buffer.alloc(ViewingKey.SIZE)),
]);
OffchainJoinSplitData.SIZE = OffchainJoinSplitData.EMPTY.toBuffer().length;
