// @aztec/barretenberg/dest/offchain_tx_data/offchain_account_data.js
import { AliasHash } from '../account_id/index.js';
import { GrumpkinAddress } from '../address/index.js';
import { numToUInt32BE } from '../serialize/index.js';
var DataSizes;
(function (DataSizes) {
    DataSizes[DataSizes["ACCOUNT_PUBLIC_KEY"] = GrumpkinAddress.SIZE] = "ACCOUNT_PUBLIC_KEY";
    DataSizes[DataSizes["ALIAS_HASH"] = AliasHash.SIZE] = "ALIAS_HASH";
    DataSizes[DataSizes["SPENDING_PUBLIC_KEY_1"] = 32] = "SPENDING_PUBLIC_KEY_1";
    DataSizes[DataSizes["SPENDING_PUBLIC_KEY_2"] = 32] = "SPENDING_PUBLIC_KEY_2";
    DataSizes[DataSizes["TX_REF_NO"] = 4] = "TX_REF_NO";
})(DataSizes || (DataSizes = {}));
var DataOffsets;
(function (DataOffsets) {
    DataOffsets[DataOffsets["ACCOUNT_PUBLIC_KEY"] = 0] = "ACCOUNT_PUBLIC_KEY";
    DataOffsets[DataOffsets["ALIAS_HASH"] = DataOffsets.ACCOUNT_PUBLIC_KEY + DataSizes.ACCOUNT_PUBLIC_KEY] = "ALIAS_HASH";
    DataOffsets[DataOffsets["SPENDING_PUBLIC_KEY_1"] = DataOffsets.ALIAS_HASH + DataSizes.ALIAS_HASH] = "SPENDING_PUBLIC_KEY_1";
    DataOffsets[DataOffsets["SPENDING_PUBLIC_KEY_2"] = DataOffsets.SPENDING_PUBLIC_KEY_1 + DataSizes.SPENDING_PUBLIC_KEY_1] = "SPENDING_PUBLIC_KEY_2";
    DataOffsets[DataOffsets["TX_REF_NO"] = DataOffsets.SPENDING_PUBLIC_KEY_2 + DataSizes.SPENDING_PUBLIC_KEY_2] = "TX_REF_NO";
})(DataOffsets || (DataOffsets = {}));
export class OffchainAccountData {
    constructor(accountPublicKey, aliasHash, spendingPublicKey1 = Buffer.alloc(32), spendingPublicKey2 = Buffer.alloc(32), txRefNo = 0) {
        this.accountPublicKey = accountPublicKey;
        this.aliasHash = aliasHash;
        this.spendingPublicKey1 = spendingPublicKey1;
        this.spendingPublicKey2 = spendingPublicKey2;
        this.txRefNo = txRefNo;
        if (spendingPublicKey1.length !== DataSizes.SPENDING_PUBLIC_KEY_1) {
            throw new Error('Expect spendingPublicKey1 to be 32 bytes.');
        }
        if (spendingPublicKey2.length !== DataSizes.SPENDING_PUBLIC_KEY_2) {
            throw new Error('Expect spendingPublicKey2 to be 32 bytes.');
        }
    }
    static fromBuffer(buf) {
        if (buf.length !== OffchainAccountData.SIZE) {
            throw new Error(`Invalid buffer size: ${buf.length} != ${OffchainAccountData.SIZE}`);
        }
        const accountPublicKey = new GrumpkinAddress(buf.slice(DataOffsets.ACCOUNT_PUBLIC_KEY, DataOffsets.ACCOUNT_PUBLIC_KEY + DataSizes.ACCOUNT_PUBLIC_KEY));
        const aliasHash = new AliasHash(buf.slice(DataOffsets.ALIAS_HASH, DataOffsets.ALIAS_HASH + DataSizes.ALIAS_HASH));
        const spendingPublicKey1 = buf.slice(DataOffsets.SPENDING_PUBLIC_KEY_1, DataOffsets.SPENDING_PUBLIC_KEY_1 + DataSizes.SPENDING_PUBLIC_KEY_1);
        const spendingPublicKey2 = buf.slice(DataOffsets.SPENDING_PUBLIC_KEY_2, DataOffsets.SPENDING_PUBLIC_KEY_2 + DataSizes.SPENDING_PUBLIC_KEY_2);
        const txRefNo = buf.readUInt32BE(DataOffsets.TX_REF_NO);
        return new OffchainAccountData(accountPublicKey, aliasHash, spendingPublicKey1, spendingPublicKey2, txRefNo);
    }
    toBuffer() {
        return Buffer.concat([
            this.accountPublicKey.toBuffer(),
            this.aliasHash.toBuffer(),
            this.spendingPublicKey1,
            this.spendingPublicKey2,
            numToUInt32BE(this.txRefNo),
        ]);
    }
}
OffchainAccountData.EMPTY = new OffchainAccountData(GrumpkinAddress.ZERO, AliasHash.ZERO);
OffchainAccountData.SIZE = OffchainAccountData.EMPTY.toBuffer().length;
