// @aztec/barretenberg/dest/account_id/account_alias_id.js
import { AliasHash } from './alias_hash.js';
export class AccountAliasId {
    constructor(aliasHash, accountNonce) {
        this.aliasHash = aliasHash;
        this.accountNonce = accountNonce;
    }
    static fromAlias(alias, accountNonce, blake2s) {
        return new AccountAliasId(AliasHash.fromAlias(alias, blake2s), accountNonce);
    }
    static random() {
        return new AccountAliasId(AliasHash.random(), 0);
    }
    static fromBuffer(id) {
        if (id.length !== 32) {
            throw new Error('Invalid id buffer.');
        }
        const aliasHash = new AliasHash(id.slice(4, 32));
        const accountNonce = id.readUInt32BE(0);
        return new AccountAliasId(aliasHash, accountNonce);
    }
    toBuffer() {
        const accountNonceBuf = Buffer.alloc(4);
        accountNonceBuf.writeUInt32BE(this.accountNonce);
        return Buffer.concat([accountNonceBuf, this.aliasHash.toBuffer()]);
    }
    toString() {
        return `0x${this.toBuffer().toString('hex')}`;
    }
    equals(rhs) {
        return this.aliasHash.equals(rhs.aliasHash) && this.accountNonce === rhs.accountNonce;
    }
}
AccountAliasId.ZERO = AccountAliasId.fromBuffer(Buffer.alloc(32));
