// @aztec/sdk/dest/transport/transport_server.js
/**
 * Keeps track of clients, providing a broadcast, and request/response api with multiplexing.
 */
export class TransportServer {
    constructor(listener, msgHandlerFn) {
        this.listener = listener;
        this.msgHandlerFn = msgHandlerFn;
        this.sockets = [];
    }
    start() {
        this.listener.on('new_socket', client => this.handleNewSocket(client));
        this.listener.open();
    }
    stop() {
        this.listener.close();
        this.sockets.forEach(socket => socket.close());
    }
    async broadcast(msg) {
        await Promise.all(this.sockets.map(s => s.send({ payload: msg })));
    }
    handleNewSocket(socket) {
        socket.registerHandler(msg => this.handleSocketMessage(socket, msg));
        this.sockets.push(socket);
    }
    async handleSocketMessage(socket, { msgId, payload }) {
        try {
            const data = await this.msgHandlerFn(payload);
            const rep = { msgId, payload: data };
            await socket.send(rep);
        }
        catch (err) {
            const rep = { msgId, error: err.message };
            await socket.send(rep);
        }
    }
}
