// @aztec/sdk/dest/transport/transport_client.js
import EventEmitter from 'events';
import { isEventMessage } from './messages.js';
import { createDebugLogger } from '@aztec/barretenberg/log';
const debug = createDebugLogger('aztec:transport_client');
/**
 * A TransportClient provides a request/response and event api to a corresponding TransportServer.
 * If `broadcast` is called on TransportServer, TransportClients will emit an `event_msg`.
 * The `request` method will block until a response is returned from the TransportServer's dispatch function.
 * Request multiplexing is supported.
 */
export class TransportClient extends EventEmitter {
    constructor(transportConnect) {
        super();
        this.transportConnect = transportConnect;
        this.msgId = 0;
        this.pendingRequests = [];
    }
    async open() {
        this.socket = await this.transportConnect.createSocket();
        this.socket.registerHandler(msg => this.handleSocketMessage(msg));
    }
    close() {
        this.socket?.close();
        this.removeAllListeners();
    }
    request(payload) {
        if (!this.socket) {
            throw new Error('Socket not open.');
        }
        const msgId = this.msgId++;
        const msg = { msgId, payload };
        debug(`->`, msg);
        return new Promise((resolve, reject) => {
            this.pendingRequests.push({ resolve, reject, msgId });
            this.socket.send(msg).catch(reject);
        });
    }
    handleSocketMessage(msg) {
        debug(`<-`, msg);
        if (isEventMessage(msg)) {
            this.emit('event_msg', msg.payload);
            return;
        }
        const reqIndex = this.pendingRequests.findIndex(r => r.msgId === msg.msgId);
        if (reqIndex === -1) {
            return;
        }
        const [pending] = this.pendingRequests.splice(reqIndex, 1);
        if (msg.error) {
            pending.reject(new Error(msg.error));
        }
        else {
            pending.resolve(msg.payload);
        }
    }
}
