// @aztec/sdk/dest/proofs/defi_deposit_proof_creator.js
import { toBigIntBE } from '@aztec/barretenberg/bigint_buffer';
import { ProofData, ProofId } from '@aztec/barretenberg/client_proofs';
import { createDebugLogger } from '@aztec/barretenberg/log';
import { OffchainDefiDepositData } from '@aztec/barretenberg/offchain_tx_data';
import { TxId } from '@aztec/barretenberg/tx_id';
import { CoreDefiTx } from '../core_tx/index.js';
import { treeNoteToNote } from '../note/index.js';
import { JoinSplitTxFactory } from './join_split_tx_factory.js';
const debug = createDebugLogger('bb:defi_deposit_proof_creator');
export class DefiDepositProofCreator {
    constructor(prover, noteAlgos, worldState, grumpkin, db) {
        this.prover = prover;
        this.noteAlgos = noteAlgos;
        this.txFactory = new JoinSplitTxFactory(noteAlgos, worldState, grumpkin, db);
    }
    async createProofInput(user, bridgeCallData, depositValue, inputNotes, spendingPublicKey) {
        const assetId = bridgeCallData.inputAssetIdA;
        const newNoteOwnerAccountRequired = !spendingPublicKey.equals(user.accountPublicKey);
        const proofInput = await this.txFactory.createTx(user, ProofId.DEFI_DEPOSIT, assetId, inputNotes, spendingPublicKey, {
            bridgeCallData,
            defiDepositValue: depositValue,
            newNoteOwner: user.accountPublicKey,
            newNoteOwnerAccountRequired,
        });
        const signingData = await this.prover.computeSigningData(proofInput.tx);
        return { ...proofInput, signingData };
    }
    async createProof({ accountPublicKey, accountPrivateKey }, { tx, signature, partialStateSecretEphPubKey, viewingKeys }, txRefNo, timeout) {
        debug('creating proof...');
        const start = new Date().getTime();
        tx.signature = signature;
        const proof = await this.prover.createProof(tx, timeout);
        debug(`created proof: ${new Date().getTime() - start}ms`);
        debug(`proof size: ${proof.length}`);
        const proofData = new ProofData(proof);
        const txId = new TxId(proofData.txId);
        const { outputNotes, claimNote: { value: depositValue, bridgeCallData, partialStateSecret }, } = tx;
        const txFee = toBigIntBE(proofData.txFee);
        const accountRequired = outputNotes[1].accountRequired;
        const partialState = this.noteAlgos.valueNotePartialCommitment(partialStateSecret, accountPublicKey, accountRequired);
        const coreTx = new CoreDefiTx(txId, accountPublicKey, bridgeCallData, depositValue, txFee, txRefNo, new Date(), partialState, partialStateSecret);
        const offchainTxData = new OffchainDefiDepositData(bridgeCallData, partialState, partialStateSecretEphPubKey, depositValue, txFee, viewingKeys[0], // contains [value, asset_id, accountRequired, creatorPubKey] of the change note (returned to the sender)
        txRefNo);
        return {
            tx: coreTx,
            proofData,
            offchainTxData,
            outputNotes: [
                treeNoteToNote(outputNotes[0], accountPrivateKey, this.noteAlgos, {
                    allowChain: proofData.allowChainFromNote1,
                }),
                treeNoteToNote(outputNotes[1], accountPrivateKey, this.noteAlgos, {
                    allowChain: proofData.allowChainFromNote2,
                }),
            ],
        };
    }
}
