// @aztec/sdk/dest/proofs/proof_input.js
import { GrumpkinAddress } from '@aztec/barretenberg/address';
import { AccountTx, JoinSplitTx, ProofData } from '@aztec/barretenberg/client_proofs';
import { SchnorrSignature } from '@aztec/barretenberg/crypto';
import { HashPath } from '@aztec/barretenberg/merkle_tree';
import { ViewingKey } from '@aztec/barretenberg/viewing_key';
import { treeNoteToNote } from '../note/index.js';
export const accountProofInputToJson = ({ tx, signingData, signature }) => ({
    tx: new Uint8Array(tx.toBuffer()),
    signingData: new Uint8Array(signingData),
    signature: signature ? signature.toString() : undefined,
});
export const accountProofInputFromJson = ({ tx, signingData, signature, }) => ({
    tx: AccountTx.fromBuffer(Buffer.from(tx)),
    signingData: Buffer.from(signingData),
    signature: signature ? SchnorrSignature.fromString(signature) : undefined,
});
export const joinSplitProofInputToJson = ({ tx, viewingKeys, partialStateSecretEphPubKey, signingData, signature, }) => ({
    tx: new Uint8Array(tx.toBuffer()),
    viewingKeys: viewingKeys.map(vk => vk.toString()),
    partialStateSecretEphPubKey: partialStateSecretEphPubKey ? partialStateSecretEphPubKey.toString() : undefined,
    signingData: new Uint8Array(signingData),
    signature: signature ? signature.toString() : undefined,
});
export const joinSplitProofInputFromJson = ({ tx, viewingKeys, partialStateSecretEphPubKey, signingData, signature, }) => ({
    tx: JoinSplitTx.fromBuffer(Buffer.from(tx)),
    viewingKeys: viewingKeys.map(vk => ViewingKey.fromString(vk)),
    partialStateSecretEphPubKey: partialStateSecretEphPubKey
        ? GrumpkinAddress.fromString(partialStateSecretEphPubKey)
        : undefined,
    signingData: Buffer.from(signingData),
    signature: signature ? SchnorrSignature.fromString(signature) : undefined,
});
export const joinSplitTxInputToJoinSplitTx = (txInput, accountPrivateKey, accountPublicKey) => new JoinSplitTx(txInput.proofId, txInput.publicValue, txInput.publicOwner, (txInput.inputNotes[0] || txInput.outputNotes[0]).assetId, txInput.inputNotes.reduce((count, n) => count + (n.value ? 1 : 0), 0), txInput.inputNotes.map(n => n.index || 0), txInput.dataRoot, txInput.inputNotes.map(n => HashPath.fromBuffer(n.hashPath)), txInput.inputNotes.map(n => n.treeNote), txInput.outputNotes.map(n => n.treeNote), txInput.claimNote, accountPrivateKey, txInput.aliasHash, !txInput.spendingPublicKey.equals(accountPublicKey), txInput.accountIndex, txInput.accountPath, txInput.spendingPublicKey, txInput.backwardLink, txInput.allowChain);
export const joinSplitTxToJoinSplitTxInput = (tx, accountPrivateKey, noteAlgos) => ({
    proofId: tx.proofId,
    publicValue: tx.publicValue,
    publicOwner: tx.publicOwner,
    inputNotes: tx.inputNotes.map((n, i) => treeNoteToNote(n, accountPrivateKey, noteAlgos, {
        gibberish: i >= tx.numInputNotes,
        index: tx.inputNoteIndices[i],
        hashPath: tx.inputNotePaths[i].toBuffer(),
    })),
    outputNotes: tx.outputNotes.map((n, i) => treeNoteToNote(n, accountPrivateKey, noteAlgos, {
        allowChain: !i ? ProofData.allowChainFromNote1(tx.allowChain) : ProofData.allowChainFromNote2(tx.allowChain),
    })),
    claimNote: tx.claimNote,
    spendingPublicKey: tx.spendingPublicKey,
    aliasHash: tx.aliasHash,
    accountIndex: tx.accountIndex,
    accountPath: tx.accountPath,
    dataRoot: tx.merkleRoot,
    backwardLink: tx.backwardLink,
    allowChain: tx.allowChain,
});
