// @aztec/sdk/dest/proofs/proof_output.js
import { ProofData, ProofId } from '@aztec/barretenberg/client_proofs';
import { OffchainAccountData, OffchainDefiDepositData, OffchainJoinSplitData, } from '@aztec/barretenberg/offchain_tx_data';
import { coreUserTxFromJson, coreUserTxToJson, } from '../core_tx/index.js';
import { noteFromJson, noteToJson } from '../note/index.js';
export const proofOutputToJson = ({ tx, proofData, offchainTxData, outputNotes, signature, }) => ({
    tx: coreUserTxToJson(tx),
    proofData: new Uint8Array(proofData.rawProofData),
    offchainTxData: new Uint8Array(offchainTxData.toBuffer()),
    outputNotes: outputNotes.map(n => noteToJson(n)),
    signature: signature ? new Uint8Array(signature) : undefined,
});
const offchainTxDataFromBuffer = (proofId, buf) => {
    switch (proofId) {
        case ProofId.DEPOSIT:
        case ProofId.WITHDRAW:
        case ProofId.SEND:
            return OffchainJoinSplitData.fromBuffer(buf);
        case ProofId.ACCOUNT:
            return OffchainAccountData.fromBuffer(buf);
        case ProofId.DEFI_DEPOSIT:
            return OffchainDefiDepositData.fromBuffer(buf);
        default:
            throw new Error(`Unsupported ProofOutput proofId: ${proofId}`);
    }
};
export const proofOutputFromJson = ({ tx, proofData, offchainTxData, outputNotes, signature, }) => ({
    tx: coreUserTxFromJson(tx),
    proofData: new ProofData(Buffer.from(proofData)),
    offchainTxData: offchainTxDataFromBuffer(tx.proofId, Buffer.from(offchainTxData)),
    outputNotes: outputNotes.map(n => noteFromJson(n)),
    signature: signature ? Buffer.from(signature) : undefined,
});
export const proofOutputToProofTx = ({ proofData, offchainTxData, signature }) => ({
    proofData: proofData.rawProofData,
    offchainTxData: offchainTxData.toBuffer(),
    depositSignature: signature,
});
