// @aztec/barretenberg/dest/client_proofs/proof_data/proof_data.js
import { createTxId } from './create_tx_id.js';
var ProofDataFields;
(function (ProofDataFields) {
    ProofDataFields[ProofDataFields["PROOF_ID"] = 0] = "PROOF_ID";
    ProofDataFields[ProofDataFields["NOTE_COMMITMENT_1"] = 1] = "NOTE_COMMITMENT_1";
    ProofDataFields[ProofDataFields["NOTE_COMMITMENT_2"] = 2] = "NOTE_COMMITMENT_2";
    ProofDataFields[ProofDataFields["NULLIFIER_1"] = 3] = "NULLIFIER_1";
    ProofDataFields[ProofDataFields["NULLIFIER_2"] = 4] = "NULLIFIER_2";
    ProofDataFields[ProofDataFields["PUBLIC_VALUE"] = 5] = "PUBLIC_VALUE";
    ProofDataFields[ProofDataFields["PUBLIC_OWNER"] = 6] = "PUBLIC_OWNER";
    ProofDataFields[ProofDataFields["PUBLIC_ASSET_ID"] = 7] = "PUBLIC_ASSET_ID";
    ProofDataFields[ProofDataFields["NOTE_TREE_ROOT"] = 8] = "NOTE_TREE_ROOT";
    ProofDataFields[ProofDataFields["TX_FEE"] = 9] = "TX_FEE";
    ProofDataFields[ProofDataFields["TX_FEE_ASSET_ID"] = 10] = "TX_FEE_ASSET_ID";
    ProofDataFields[ProofDataFields["BRIDGE_CALL_DATA"] = 11] = "BRIDGE_CALL_DATA";
    ProofDataFields[ProofDataFields["DEFI_DEPOSIT_VALUE"] = 12] = "DEFI_DEPOSIT_VALUE";
    ProofDataFields[ProofDataFields["DEFI_ROOT"] = 13] = "DEFI_ROOT";
    ProofDataFields[ProofDataFields["BACKWARD_LINK"] = 14] = "BACKWARD_LINK";
    ProofDataFields[ProofDataFields["ALLOW_CHAIN"] = 15] = "ALLOW_CHAIN";
})(ProofDataFields || (ProofDataFields = {}));
var ProofDataOffsets;
(function (ProofDataOffsets) {
    ProofDataOffsets[ProofDataOffsets["PROOF_ID"] = 28] = "PROOF_ID";
    ProofDataOffsets[ProofDataOffsets["NOTE_COMMITMENT_1"] = 32] = "NOTE_COMMITMENT_1";
    ProofDataOffsets[ProofDataOffsets["NOTE_COMMITMENT_2"] = 64] = "NOTE_COMMITMENT_2";
    ProofDataOffsets[ProofDataOffsets["NULLIFIER_1"] = 96] = "NULLIFIER_1";
    ProofDataOffsets[ProofDataOffsets["NULLIFIER_2"] = 128] = "NULLIFIER_2";
    ProofDataOffsets[ProofDataOffsets["PUBLIC_VALUE"] = 160] = "PUBLIC_VALUE";
    ProofDataOffsets[ProofDataOffsets["PUBLIC_OWNER"] = 192] = "PUBLIC_OWNER";
    ProofDataOffsets[ProofDataOffsets["PUBLIC_ASSET_ID"] = 224] = "PUBLIC_ASSET_ID";
    ProofDataOffsets[ProofDataOffsets["NOTE_TREE_ROOT"] = 256] = "NOTE_TREE_ROOT";
    ProofDataOffsets[ProofDataOffsets["TX_FEE"] = 288] = "TX_FEE";
    ProofDataOffsets[ProofDataOffsets["TX_FEE_ASSET_ID"] = 320] = "TX_FEE_ASSET_ID";
    ProofDataOffsets[ProofDataOffsets["BRIDGE_CALL_DATA"] = 352] = "BRIDGE_CALL_DATA";
    ProofDataOffsets[ProofDataOffsets["DEFI_DEPOSIT_VALUE"] = 384] = "DEFI_DEPOSIT_VALUE";
    ProofDataOffsets[ProofDataOffsets["DEFI_ROOT"] = 416] = "DEFI_ROOT";
    ProofDataOffsets[ProofDataOffsets["BACKWARD_LINK"] = 448] = "BACKWARD_LINK";
    ProofDataOffsets[ProofDataOffsets["ALLOW_CHAIN"] = 480] = "ALLOW_CHAIN";
})(ProofDataOffsets || (ProofDataOffsets = {}));
/**
 * Represents tx proof data as returned by the proof generator.
 * Differs to on chain data, in that not all data here is actually published.
 * Fields that differ between proofs, or natural buffers, are of type Buffer.
 * Fields that are always of fixed type/meaning are converted.
 */
export class ProofData {
    constructor(rawProofData, txId) {
        this.rawProofData = rawProofData;
        this.proofId = rawProofData.readUInt32BE(ProofDataOffsets.PROOF_ID);
        this.noteCommitment1 = rawProofData.slice(ProofDataOffsets.NOTE_COMMITMENT_1, ProofDataOffsets.NOTE_COMMITMENT_1 + 32);
        this.noteCommitment2 = rawProofData.slice(ProofDataOffsets.NOTE_COMMITMENT_2, ProofDataOffsets.NOTE_COMMITMENT_2 + 32);
        this.nullifier1 = rawProofData.slice(ProofDataOffsets.NULLIFIER_1, ProofDataOffsets.NULLIFIER_1 + 32);
        this.nullifier2 = rawProofData.slice(ProofDataOffsets.NULLIFIER_2, ProofDataOffsets.NULLIFIER_2 + 32);
        this.publicValue = rawProofData.slice(ProofDataOffsets.PUBLIC_VALUE, ProofDataOffsets.PUBLIC_VALUE + 32);
        this.publicOwner = rawProofData.slice(ProofDataOffsets.PUBLIC_OWNER, ProofDataOffsets.PUBLIC_OWNER + 32);
        this.publicAssetId = rawProofData.slice(ProofDataOffsets.PUBLIC_ASSET_ID, ProofDataOffsets.PUBLIC_ASSET_ID + 32);
        // Not published as part of inner proofs.
        this.noteTreeRoot = rawProofData.slice(ProofDataOffsets.NOTE_TREE_ROOT, ProofDataOffsets.NOTE_TREE_ROOT + 32);
        this.txFee = rawProofData.slice(ProofDataOffsets.TX_FEE, ProofDataOffsets.TX_FEE + 32);
        this.txFeeAssetId = rawProofData.slice(ProofDataOffsets.TX_FEE_ASSET_ID, ProofDataOffsets.TX_FEE_ASSET_ID + 32);
        this.bridgeCallData = rawProofData.slice(ProofDataOffsets.BRIDGE_CALL_DATA, ProofDataOffsets.BRIDGE_CALL_DATA + 32);
        this.defiDepositValue = rawProofData.slice(ProofDataOffsets.DEFI_DEPOSIT_VALUE, ProofDataOffsets.DEFI_DEPOSIT_VALUE + 32);
        this.defiRoot = rawProofData.slice(ProofDataOffsets.DEFI_ROOT, ProofDataOffsets.DEFI_ROOT + 32);
        this.backwardLink = rawProofData.slice(ProofDataOffsets.BACKWARD_LINK, ProofDataOffsets.BACKWARD_LINK + 32);
        this.allowChain = rawProofData.slice(ProofDataOffsets.ALLOW_CHAIN, ProofDataOffsets.ALLOW_CHAIN + 32);
        this.txId_ = txId;
    }
    static getProofIdFromBuffer(rawProofData) {
        return rawProofData.readUInt32BE(ProofDataOffsets.PROOF_ID);
    }
    static allowChainFromNote1(allowChain) {
        return [1, 3].includes(allowChain);
    }
    static allowChainFromNote2(allowChain) {
        return [2, 3].includes(allowChain);
    }
    get allowChainFromNote1() {
        const allowChain = this.allowChain.readUInt32BE(28);
        return ProofData.allowChainFromNote1(allowChain);
    }
    get allowChainFromNote2() {
        const allowChain = this.allowChain.readUInt32BE(28);
        return ProofData.allowChainFromNote2(allowChain);
    }
    get feeAssetId() {
        return this.txFeeAssetId.readUInt32BE(28);
    }
    get txId() {
        if (!this.txId_) {
            this.txId_ = createTxId(this.rawProofData.slice(0, ProofData.NUM_PUBLISHED_PUBLIC_INPUTS * 32));
        }
        return this.txId_;
    }
}
ProofData.NUM_PUBLIC_INPUTS = 17;
ProofData.NUM_PUBLISHED_PUBLIC_INPUTS = 8;
