// @aztec/barretenberg/dest/rollup_provider/rollup_provider.js
import { EthAddress } from '../address/index.js';
import { TxId } from '../tx_id/index.js';
export var TxSettlementTime;
(function (TxSettlementTime) {
    TxSettlementTime[TxSettlementTime["NEXT_ROLLUP"] = 0] = "NEXT_ROLLUP";
    TxSettlementTime[TxSettlementTime["INSTANT"] = 1] = "INSTANT";
})(TxSettlementTime = TxSettlementTime || (TxSettlementTime = {}));
export var DefiSettlementTime;
(function (DefiSettlementTime) {
    DefiSettlementTime[DefiSettlementTime["DEADLINE"] = 0] = "DEADLINE";
    DefiSettlementTime[DefiSettlementTime["NEXT_ROLLUP"] = 1] = "NEXT_ROLLUP";
    DefiSettlementTime[DefiSettlementTime["INSTANT"] = 2] = "INSTANT";
})(DefiSettlementTime = DefiSettlementTime || (DefiSettlementTime = {}));
export const txToJson = ({ proofData, offchainTxData, depositSignature }) => ({
    proofData: proofData.toString('hex'),
    offchainTxData: offchainTxData.toString('hex'),
    depositSignature: depositSignature ? depositSignature.toString('hex') : undefined,
});
export const txFromJson = ({ proofData, offchainTxData, depositSignature }) => ({
    proofData: Buffer.from(proofData, 'hex'),
    offchainTxData: Buffer.from(offchainTxData, 'hex'),
    depositSignature: depositSignature ? Buffer.from(depositSignature, 'hex') : undefined,
});
export const pendingTxToJson = ({ txId, noteCommitment1, noteCommitment2 }) => ({
    txId: txId.toString(),
    noteCommitment1: noteCommitment1.toString('hex'),
    noteCommitment2: noteCommitment2.toString('hex'),
});
export const pendingTxFromJson = ({ txId, noteCommitment1, noteCommitment2 }) => ({
    txId: TxId.fromString(txId),
    noteCommitment1: Buffer.from(noteCommitment1, 'hex'),
    noteCommitment2: Buffer.from(noteCommitment2, 'hex'),
});
export const initialWorldStateToBuffer = (initialWorldState) => {
    const accountsSizeBuf = Buffer.alloc(4);
    accountsSizeBuf.writeUInt32BE(initialWorldState.initialAccounts.length);
    return Buffer.concat([accountsSizeBuf, initialWorldState.initialAccounts, ...initialWorldState.initialSubtreeRoots]);
};
export const initialWorldStateFromBuffer = (data) => {
    const accountsSize = data.readUInt32BE(0);
    const subTreeStart = 4 + accountsSize;
    const initialWorldState = {
        initialAccounts: data.slice(4, subTreeStart),
        initialSubtreeRoots: [],
    };
    // each sub tree root is 32 bytes
    for (let i = subTreeStart; i < data.length; i += 32) {
        initialWorldState.initialSubtreeRoots.push(data.slice(i, i + 32));
    }
    return initialWorldState;
};
export const depositTxToJson = ({ assetId, value, publicOwner }) => ({
    assetId,
    value: value.toString(),
    publicOwner: publicOwner.toString(),
});
export const depositTxFromJson = ({ assetId, value, publicOwner }) => ({
    assetId,
    value: BigInt(value),
    publicOwner: EthAddress.fromString(publicOwner),
});
