// @aztec/barretenberg/dest/rollup_provider/rollup_provider_status.js
import { EthAddress } from '../address/index.js';
import { blockchainStatusFromJson, blockchainStatusToJson, } from '../blockchain/index.js';
import { bridgeStatusFromJson, bridgeStatusToJson } from './bridge_status.js';
import { privacySetsFromJson, privacySetsToJson } from './privacy_set.js';
export * from './bridge_config.js';
export * from './bridge_status.js';
export * from './privacy_set.js';
export const runtimeConfigToJson = ({ maxFeeGasPrice, maxFeePerGas, maxPriorityFeePerGas, privacySets, rollupBeneficiary, blacklist, ...rest }) => ({
    ...rest,
    maxFeeGasPrice: maxFeeGasPrice.toString(),
    maxFeePerGas: maxFeePerGas.toString(),
    maxPriorityFeePerGas: maxPriorityFeePerGas.toString(),
    privacySets: privacySetsToJson(privacySets),
    rollupBeneficiary: rollupBeneficiary ? rollupBeneficiary.toLowerCaseAddress() : undefined,
    blacklist: blacklist ? blacklist.map(x => x.toLowerCaseAddress()) : undefined,
});
export const runtimeConfigFromJson = ({ maxFeeGasPrice, maxFeePerGas, maxPriorityFeePerGas, privacySets, rollupBeneficiary, blacklist, ...rest }) => ({
    ...rest,
    maxFeeGasPrice: BigInt(maxFeeGasPrice),
    maxFeePerGas: BigInt(maxFeePerGas),
    maxPriorityFeePerGas: BigInt(maxPriorityFeePerGas),
    privacySets: privacySetsFromJson(privacySets),
    rollupBeneficiary: rollupBeneficiary ? EthAddress.fromString(rollupBeneficiary) : undefined,
    blacklist: blacklist ? blacklist.map(x => EthAddress.fromString(x)) : undefined,
});
export const partialRuntimeConfigFromJson = ({ maxFeeGasPrice, maxFeePerGas, maxPriorityFeePerGas, privacySets, rollupBeneficiary, blacklist, ...rest }) => ({
    ...rest,
    ...(maxFeeGasPrice !== undefined ? { maxFeeGasPrice: BigInt(maxFeeGasPrice) } : {}),
    ...(maxFeePerGas !== undefined ? { maxFeePerGas: BigInt(maxFeePerGas) } : {}),
    ...(maxPriorityFeePerGas !== undefined ? { maxPriorityFeePerGas: BigInt(maxPriorityFeePerGas) } : {}),
    ...(privacySets ? { privacySets: privacySetsFromJson(privacySets) } : {}),
    ...(rollupBeneficiary ? { rollupBeneficiary: EthAddress.fromString(rollupBeneficiary) } : {}),
    ...(blacklist ? { blacklist: blacklist.map(x => EthAddress.fromString(x)) } : {}),
});
export const rollupProviderStatusToJson = ({ blockchainStatus, nextPublishTime, runtimeConfig, bridgeStatus, ...rest }) => ({
    ...rest,
    blockchainStatus: blockchainStatusToJson(blockchainStatus),
    nextPublishTime: nextPublishTime.toISOString(),
    runtimeConfig: runtimeConfigToJson(runtimeConfig),
    bridgeStatus: bridgeStatus.map(bridgeStatusToJson),
});
export const rollupProviderStatusFromJson = ({ blockchainStatus, nextPublishTime, runtimeConfig, bridgeStatus, ...rest }) => ({
    ...rest,
    blockchainStatus: blockchainStatusFromJson(blockchainStatus),
    nextPublishTime: new Date(nextPublishTime),
    runtimeConfig: runtimeConfigFromJson(runtimeConfig),
    bridgeStatus: bridgeStatus.map(bridgeStatusFromJson),
});
