// @aztec/blockchain/dest/contracts/fee_distributor/fee_distributor.js
import { EthAddress } from '@aztec/barretenberg/address';
import { TxHash } from '@aztec/barretenberg/blockchain';
import { Web3Provider } from '@ethersproject/providers';
import { Contract } from 'ethers';
import { AztecFeeDistributor } from '../../abis.js';
const fixEthersStackTrace = (err) => {
    err.stack += new Error().stack;
    throw err;
};
export class FeeDistributor {
    constructor(feeDistributorContractAddress, provider, defaults = {}) {
        this.feeDistributorContractAddress = feeDistributorContractAddress;
        this.provider = provider;
        this.defaults = defaults;
        this.feeDistributor = new Contract(feeDistributorContractAddress.toString(), AztecFeeDistributor.abi, new Web3Provider(this.provider));
    }
    get address() {
        return this.feeDistributorContractAddress;
    }
    get contract() {
        return this.feeDistributor;
    }
    async WETH() {
        return EthAddress.fromString(await this.feeDistributor.WETH().catch(fixEthersStackTrace));
    }
    async aztecFeeClaimer() {
        return EthAddress.fromString(await this.feeDistributor.aztecFeeClaimer().catch(fixEthersStackTrace));
    }
    async feeLimit() {
        return BigInt(await this.feeDistributor.feeLimit().catch(fixEthersStackTrace));
    }
    async convertConstant() {
        return BigInt(await this.feeDistributor.convertConstant().catch(fixEthersStackTrace));
    }
    async txFeeBalance(asset) {
        return BigInt(await this.feeDistributor.txFeeBalance(asset.toString()).catch(fixEthersStackTrace));
    }
    async convert(asset, minOutputValue, options = this.defaults) {
        const { gasLimit } = options;
        const contract = this.getContractWithSigner(options);
        const tx = await contract.convert(asset.toString(), minOutputValue, { gasLimit }).catch(fixEthersStackTrace);
        return TxHash.fromString(tx.hash);
    }
    async setConvertConstant(constant, options = this.defaults) {
        const { gasLimit } = options;
        const contract = this.getContractWithSigner(options);
        const tx = await contract.setConvertConstant(constant, { gasLimit }).catch(fixEthersStackTrace);
        return TxHash.fromString(tx.hash);
    }
    async setFeeLimit(constant, options = this.defaults) {
        const { gasLimit } = options;
        const contract = this.getContractWithSigner(options);
        const tx = await contract.setFeeLimit(constant, { gasLimit }).catch(fixEthersStackTrace);
        return TxHash.fromString(tx.hash);
    }
    async setFeeClaimer(address, options = this.defaults) {
        const { gasLimit } = options;
        const contract = this.getContractWithSigner(options);
        const tx = await contract.setFeeClaimer(address.toString(), { gasLimit }).catch(fixEthersStackTrace);
        return TxHash.fromString(tx.hash);
    }
    async getLastReimbursement() {
        const eventFilter = this.feeDistributor.filters.FeeReimbursed();
        const events = await this.feeDistributor.queryFilter(eventFilter);
        const lastEvent = events[events.length - 1];
        return BigInt(lastEvent.args.amount);
    }
    getContractWithSigner(options) {
        const { provider = this.provider, signingAddress } = options;
        const ethSigner = new Web3Provider(provider).getSigner(signingAddress ? signingAddress.toString() : 0);
        return new Contract(this.feeDistributorContractAddress.toString(), AztecFeeDistributor.abi, ethSigner);
    }
}
