// @aztec/barretenberg/dest/rollup_proof/inner_proof.js
import { createTxId, ProofId } from '../client_proofs/index.js';
import { numToUInt32BE } from '../serialize/index.js';
export class InnerProofData {
    constructor(proofId, noteCommitment1, noteCommitment2, nullifier1, nullifier2, publicValue, publicOwner, publicAssetId) {
        this.proofId = proofId;
        this.noteCommitment1 = noteCommitment1;
        this.noteCommitment2 = noteCommitment2;
        this.nullifier1 = nullifier1;
        this.nullifier2 = nullifier2;
        this.publicValue = publicValue;
        this.publicOwner = publicOwner;
        this.publicAssetId = publicAssetId;
    }
    get txId() {
        if (!this.txId_) {
            this.txId_ = createTxId(this.toBuffer());
        }
        return this.txId_;
    }
    getDepositSigningData() {
        return this.toBuffer();
    }
    toBuffer() {
        return Buffer.concat([
            numToUInt32BE(this.proofId, 32),
            this.noteCommitment1,
            this.noteCommitment2,
            this.nullifier1,
            this.nullifier2,
            this.publicValue,
            this.publicOwner,
            this.publicAssetId,
        ]);
    }
    isPadding() {
        return this.proofId === ProofId.PADDING;
    }
    static fromBuffer(innerPublicInputs) {
        let dataStart = 0;
        const proofId = innerPublicInputs.readUInt32BE(dataStart + 28);
        dataStart += 32;
        const noteCommitment1 = innerPublicInputs.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const noteCommitment2 = innerPublicInputs.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const nullifier1 = innerPublicInputs.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const nullifier2 = innerPublicInputs.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const publicValue = innerPublicInputs.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const publicOwner = innerPublicInputs.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const publicAssetId = innerPublicInputs.slice(dataStart, dataStart + 32);
        dataStart += 32;
        return new InnerProofData(proofId, noteCommitment1, noteCommitment2, nullifier1, nullifier2, publicValue, publicOwner, publicAssetId);
    }
}
InnerProofData.NUM_PUBLIC_INPUTS = 8;
InnerProofData.LENGTH = InnerProofData.NUM_PUBLIC_INPUTS * 32;
InnerProofData.PADDING = InnerProofData.fromBuffer(Buffer.alloc(InnerProofData.LENGTH));
