// @aztec/sdk/dest/proofs/proof_input/proof_input.js
import { GrumpkinAddress } from '@aztec/barretenberg/address';
import { AccountTx } from '@aztec/barretenberg/client_proofs';
import { ViewingKey } from '@aztec/barretenberg/viewing_key';
import { joinSplitTxInputFromJson, joinSplitTxInputToJson, } from './join_split_tx_input.js';
export { AccountTx } from '@aztec/barretenberg/client_proofs';
export const accountProofInputToJson = ({ tx, signingData }) => ({
    tx: new Uint8Array(tx.toBuffer()),
    signingData: new Uint8Array(signingData),
});
export const accountProofInputFromJson = ({ tx, signingData }) => ({
    tx: AccountTx.fromBuffer(Buffer.from(tx)),
    signingData: Buffer.from(signingData),
});
export const paymentProofInputToJson = ({ tx, viewingKeys, signingData, }) => ({
    tx: joinSplitTxInputToJson(tx),
    viewingKeys: viewingKeys.map(vk => vk.toString()),
    signingData: new Uint8Array(signingData),
});
export const paymentProofInputFromJson = ({ tx, viewingKeys, signingData, }) => ({
    tx: joinSplitTxInputFromJson(tx),
    viewingKeys: viewingKeys.map(vk => ViewingKey.fromString(vk)),
    signingData: Buffer.from(signingData),
});
export const defiProofInputToJson = ({ tx, viewingKey, partialStateSecretEphPubKey, signingData, }) => ({
    tx: joinSplitTxInputToJson(tx),
    viewingKey: viewingKey.toString(),
    partialStateSecretEphPubKey: partialStateSecretEphPubKey.toString(),
    signingData: new Uint8Array(signingData),
});
export const defiProofInputFromJson = ({ tx, viewingKey, partialStateSecretEphPubKey, signingData, }) => ({
    tx: joinSplitTxInputFromJson(tx),
    viewingKey: ViewingKey.fromString(viewingKey),
    partialStateSecretEphPubKey: GrumpkinAddress.fromString(partialStateSecretEphPubKey),
    signingData: Buffer.from(signingData),
});
