// @aztec/barretenberg/dest/pippenger/single_pippenger.js
import { Transfer } from '../transport/index.js';
export class SinglePippenger {
    constructor(wasm) {
        this.wasm = wasm;
    }
    async init(crsData) {
        this.numPoints = crsData.length / 64;
        // The allocation is as per the point_table_size in pippenger.hpp.
        // The crs data does not have the affine_one point at the start.
        // affine_one is filled in at the first position by new_pippenger before building the point table.
        // The last point is discarded, so we still end up with numPoints points.
        const crsPtr = await this.wasm.call('bbmalloc', 64 * (this.numPoints * 2 + 16));
        await this.wasm.transferToHeap(crsData.slice(0, -64), crsPtr + 64);
        this.pippengerPtr = await this.wasm.call('new_pippenger', crsPtr, this.numPoints);
    }
    async destroy() {
        await this.wasm.call('delete_pippenger', this.pippengerPtr);
    }
    async pippengerUnsafe(scalars, from, range) {
        const mem = await this.wasm.call('bbmalloc', scalars.length);
        await this.wasm.transferToHeap(Transfer(scalars, [scalars.buffer]), mem);
        await this.wasm.call('pippenger_unsafe', this.pippengerPtr, mem, from, range, 0);
        await this.wasm.call('bbfree', mem);
        return Buffer.from(await this.wasm.sliceMemory(0, 96));
    }
    async sumElements(buffer) {
        const mem = await this.wasm.call('bbmalloc', buffer.length);
        await this.wasm.transferToHeap(buffer, mem);
        await this.wasm.call('g1_sum', mem, buffer.length / 96, 0);
        await this.wasm.call('bbfree', mem);
        return Buffer.from(await this.wasm.sliceMemory(0, 96));
    }
    getPointer() {
        return this.pippengerPtr;
    }
    getWorker() {
        return this.wasm;
    }
}
