// @aztec/sdk/dest/core_sdk/sdk_status.js
import { EthAddress } from '@aztec/barretenberg/address';
export var SdkEvent;
(function (SdkEvent) {
    // The SDK version does not match the rollup provider version
    SdkEvent["VERSION_MISMATCH"] = "SDKEVENT_VERSION_MISMATCH";
    // A users state has changed.
    SdkEvent["UPDATED_USER_STATE"] = "SDKEVENT_UPDATED_USER_STATE";
    // The world state has updated. Used for displaying sync progress.
    SdkEvent["UPDATED_WORLD_STATE"] = "SDKEVENT_UPDATED_WORLD_STATE";
    // The sdk has been destroyed.
    SdkEvent["DESTROYED"] = "SDKEVENT_DESTROYED";
})(SdkEvent = SdkEvent || (SdkEvent = {}));
export const sdkStatusToJson = (status) => ({
    ...status,
    rollupContractAddress: status.rollupContractAddress.toString(),
    permitHelperContractAddress: status.permitHelperContractAddress.toString(),
    verifierContractAddress: status.verifierContractAddress.toString(),
    dataRoot: status.dataRoot.toString('hex'),
});
export const sdkStatusFromJson = (json) => ({
    ...json,
    rollupContractAddress: EthAddress.fromString(json.rollupContractAddress),
    permitHelperContractAddress: EthAddress.fromString(json.permitHelperContractAddress),
    verifierContractAddress: EthAddress.fromString(json.verifierContractAddress),
    dataRoot: Buffer.from(json.dataRoot, 'hex'),
});
