// @aztec/sdk/dest/controllers/add_spending_key_controller.js
import { proofOutputToProofTx } from '../proofs/index.js';
import { createTxRefNo } from './create_tx_ref_no.js';
export class AddSpendingKeyController {
    constructor(userId, userSigner, spendingPublicKey1, spendingPublicKey2, fee, core) {
        this.userId = userId;
        this.userSigner = userSigner;
        this.spendingPublicKey1 = spendingPublicKey1;
        this.spendingPublicKey2 = spendingPublicKey2;
        this.fee = fee;
        this.core = core;
        this.feeProofOutputs = [];
        this.txIds = [];
        this.requireFeePayingTx = !!fee.value;
    }
    async createProof(timeout) {
        const txRefNo = this.requireFeePayingTx ? createTxRefNo() : 0;
        const spendingPublicKey = this.userSigner.getPublicKey();
        if (this.requireFeePayingTx) {
            const spendingKeyRequired = !spendingPublicKey.equals(this.userId);
            const feeProofInputs = await this.core.createPaymentProofInputs(this.userId, this.fee.assetId, BigInt(0), BigInt(0), this.fee.value, BigInt(0), BigInt(0), this.userId, spendingKeyRequired, undefined, spendingPublicKey, 2);
            this.feeProofOutputs = [];
            for (const proofInput of feeProofInputs) {
                proofInput.signature = await this.userSigner.signMessage(proofInput.signingData);
                this.feeProofOutputs.push(await this.core.createPaymentProof(proofInput, txRefNo, timeout));
            }
        }
        {
            const proofInput = await this.core.createAccountProofInput(this.userId, spendingPublicKey, false, undefined, this.spendingPublicKey1, this.spendingPublicKey2, undefined);
            proofInput.signature = await this.userSigner.signMessage(proofInput.signingData);
            this.proofOutput = await this.core.createAccountProof(proofInput, txRefNo, timeout);
        }
    }
    exportProofTxs() {
        if (!this.proofOutput) {
            throw new Error('Call createProof() first.');
        }
        return [this.proofOutput, ...this.feeProofOutputs].map(proofOutputToProofTx);
    }
    async send() {
        if (!this.proofOutput) {
            throw new Error('Call createProof() first.');
        }
        this.txIds = await this.core.sendProofs([this.proofOutput, ...this.feeProofOutputs]);
        return this.txIds[0];
    }
    getTxIds() {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        return this.txIds;
    }
    async awaitSettlement(timeout) {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        await Promise.all(this.txIds.map(txId => this.core.awaitSettlement(txId, timeout)));
    }
}
