// @aztec/sdk/dest/controllers/defi_controller.js
import { validateBridgeCallData } from '@aztec/barretenberg/bridge_call_data';
import { proofOutputToProofTx } from '../proofs/index.js';
import { createTxRefNo } from './create_tx_ref_no.js';
export class DefiController {
    constructor(userId, userSigner, bridgeCallData, assetValue, fee, core) {
        this.userId = userId;
        this.userSigner = userSigner;
        this.bridgeCallData = bridgeCallData;
        this.assetValue = assetValue;
        this.fee = fee;
        this.core = core;
        this.jsProofOutputs = [];
        this.feeProofOutputs = [];
        this.txIds = [];
        if (!assetValue.value) {
            throw new Error('Deposit value must be greater than 0.');
        }
        if (assetValue.assetId !== bridgeCallData.inputAssetIdA &&
            (bridgeCallData.inputAssetIdB === undefined || assetValue.assetId !== bridgeCallData.inputAssetIdB)) {
            throw new Error(`Incorrect deposit asset. Expect ${bridgeCallData.inputAssetIdA}${bridgeCallData.inputAssetIdB !== undefined ? ` or ${bridgeCallData.inputAssetIdB}` : ''}. Got ${assetValue.assetId}.`);
        }
        validateBridgeCallData(bridgeCallData);
        if (fee.value && fee.assetId === bridgeCallData.inputAssetIdB) {
            throw new Error('Fee paying asset must be the first input asset.');
        }
        this.requireFeePayingTx = !!fee.value && fee.assetId !== bridgeCallData.inputAssetIdA;
    }
    async createProof(timeout) {
        const { value } = this.assetValue;
        const spendingPublicKey = this.userSigner.getPublicKey();
        const spendingKeyRequired = !spendingPublicKey.equals(this.userId);
        const proofInputs = await this.core.createDefiProofInput(this.userId, this.bridgeCallData, value, this.requireFeePayingTx ? BigInt(0) : this.fee.value, spendingPublicKey);
        for (const proofInput of proofInputs) {
            proofInput.signature = await this.userSigner.signMessage(proofInput.signingData);
        }
        const txRefNo = this.requireFeePayingTx || proofInputs.length > 1 ? createTxRefNo() : 0;
        const joinSplitProofInputs = proofInputs.slice(0, -1);
        const defiProofInput = proofInputs[proofInputs.length - 1];
        for (const proofInput of joinSplitProofInputs) {
            this.jsProofOutputs.push(await this.core.createPaymentProof(proofInput, txRefNo, timeout));
        }
        this.proofOutput = await this.core.createDefiProof(defiProofInput, txRefNo, timeout);
        if (this.requireFeePayingTx) {
            const feeProofInputs = await this.core.createPaymentProofInputs(this.userId, this.fee.assetId, BigInt(0), BigInt(0), this.fee.value, BigInt(0), BigInt(0), this.userId, spendingKeyRequired, undefined, spendingPublicKey, 2);
            this.feeProofOutputs = [];
            for (const proofInput of feeProofInputs) {
                proofInput.signature = await this.userSigner.signMessage(proofInput.signingData);
                this.feeProofOutputs.push(await this.core.createPaymentProof(proofInput, txRefNo, timeout));
            }
        }
    }
    exportProofTxs() {
        if (!this.proofOutput) {
            throw new Error('Call createProof() first.');
        }
        return [...this.jsProofOutputs, this.proofOutput, ...this.feeProofOutputs].map(proofOutputToProofTx);
    }
    async send() {
        if (!this.proofOutput) {
            throw new Error('Call createProof() first.');
        }
        this.txIds = await this.core.sendProofs([...this.jsProofOutputs, this.proofOutput, ...this.feeProofOutputs]);
        return this.getDefiTxId();
    }
    getTxIds() {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        return this.txIds;
    }
    async awaitDefiDepositCompletion(timeout) {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        await Promise.all(this.txIds.map(txId => this.core.awaitSettlement(txId, timeout)));
    }
    async awaitDefiFinalisation(timeout) {
        const txId = this.getDefiTxId();
        if (!txId) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        await this.core.awaitDefiFinalisation(txId, timeout);
    }
    async awaitSettlement(timeout) {
        const txId = this.getDefiTxId();
        if (!txId) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        await this.core.awaitDefiSettlement(txId, timeout);
    }
    getInteractionNonce() {
        const txId = this.getDefiTxId();
        if (!txId) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        return this.core.getDefiInteractionNonce(txId);
    }
    getDefiTxId() {
        return this.txIds[this.jsProofOutputs.length];
    }
}
