// @aztec/sdk/dest/controllers/deposit_controller.js
import { proofOutputToProofTx } from '../proofs/index.js';
import { DepositHandler } from './deposit_handler.js';
export class DepositController extends DepositHandler {
    constructor(assetValue, fee, depositor, recipient, recipientSpendingKeyRequired, core, blockchain, provider) {
        super(assetValue, fee, depositor, recipient, recipientSpendingKeyRequired, core, blockchain, provider);
        this.assetValue = assetValue;
        this.fee = fee;
        this.depositor = depositor;
        this.recipient = recipient;
        this.recipientSpendingKeyRequired = recipientSpendingKeyRequired;
        this.core = core;
        this.txIds = [];
    }
    async createProof(timeout) {
        // txRefNo is not required for creating a single deposit proof.
        await super.createProof(0, timeout);
    }
    exportProofTxs() {
        return [super.getProofOutput()].map(proofOutputToProofTx);
    }
    async send() {
        this.txIds = await this.core.sendProofs([super.getProofOutput()]);
        return this.txIds[0];
    }
    getTxIds() {
        if (!this.txIds.length) {
            throw new Error('Call send() first.');
        }
        return this.txIds;
    }
    async awaitSettlement(timeout) {
        if (!this.txIds.length) {
            throw new Error('Call send() first.');
        }
        await Promise.all(this.txIds.map(txId => this.core.awaitSettlement(txId, timeout)));
    }
}
