// @aztec/sdk/dest/controllers/fee_controller.js
import { proofOutputToProofTx } from '../proofs/index.js';
import { createTxRefNo } from './create_tx_ref_no.js';
export class FeeController {
    constructor(userId, userSigner, proofTxs, fee, core) {
        this.userId = userId;
        this.userSigner = userSigner;
        this.proofTxs = proofTxs;
        this.fee = fee;
        this.core = core;
        this.feeProofOutputs = [];
        this.txIds = [];
        if (!fee.value) {
            throw new Error('Fee cannot be 0.');
        }
    }
    async createProof(timeout) {
        const spendingPublicKey = this.userSigner.getPublicKey();
        const spendingKeyRequired = !spendingPublicKey.equals(this.userId);
        const feeProofInputs = await this.core.createPaymentProofInputs(this.userId, this.fee.assetId, BigInt(0), BigInt(0), this.fee.value, BigInt(0), BigInt(0), this.userId, spendingKeyRequired, undefined, spendingPublicKey, 2);
        const txRefNo = feeProofInputs.length > 1 ? createTxRefNo() : 0;
        this.feeProofOutputs = [];
        for (const proofInput of feeProofInputs) {
            proofInput.signature = await this.userSigner.signMessage(proofInput.signingData);
            this.feeProofOutputs.push(await this.core.createPaymentProof(proofInput, txRefNo, timeout));
        }
    }
    exportProofTxs() {
        if (!this.feeProofOutputs.length) {
            throw new Error('Call createProof() first.');
        }
        return this.feeProofOutputs.map(proofOutputToProofTx);
    }
    async send() {
        if (!this.feeProofOutputs.length) {
            throw new Error('Call createProof() first.');
        }
        this.txIds = await this.core.sendProofs(this.feeProofOutputs, this.proofTxs);
        return this.txIds[this.feeProofOutputs.length - 1];
    }
    getTxIds() {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.feeProofOutputs.length ? 'createProof()' : 'send()'} first.`);
        }
        return this.txIds;
    }
    async awaitSettlement(timeout) {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.feeProofOutputs.length ? 'createProof()' : 'send()'} first.`);
        }
        await Promise.all(this.txIds.map(txId => this.core.awaitSettlement(txId, timeout)));
    }
}
