// @aztec/sdk/dest/controllers/migrate_account_controller.js
import { EthAddress } from '@aztec/barretenberg/address';
import { proofOutputToProofTx } from '../proofs/index.js';
import { createTxRefNo } from './create_tx_ref_no.js';
import { DepositHandler } from './deposit_handler.js';
export class MigrateAccountController extends DepositHandler {
    constructor(userId, userSigner, newAccountPrivateKey, newSpendingPublicKey, recoveryPublicKey, deposit, fee, depositor = EthAddress.ZERO, core, blockchain, provider) {
        super(deposit, fee, depositor, userId, true, core, blockchain, provider);
        this.userId = userId;
        this.userSigner = userSigner;
        this.newAccountPrivateKey = newAccountPrivateKey;
        this.newSpendingPublicKey = newSpendingPublicKey;
        this.recoveryPublicKey = recoveryPublicKey;
        this.deposit = deposit;
        this.fee = fee;
        this.depositor = depositor;
        this.core = core;
        this.txIds = [];
        this.requireDeposit = !!this.publicInput.value;
    }
    async createProof(timeout) {
        const txRefNo = this.requireDeposit ? createTxRefNo() : 0;
        if (this.requireDeposit) {
            await super.createProof(txRefNo, timeout);
        }
        const spendingPublicKey = this.userSigner.getPublicKey();
        const proofInput = await this.core.createAccountProofInput(this.userId, spendingPublicKey, true, undefined, this.newSpendingPublicKey, this.recoveryPublicKey, this.newAccountPrivateKey);
        proofInput.signature = await this.userSigner.signMessage(proofInput.signingData);
        this.proofOutput = await this.core.createAccountProof(proofInput, txRefNo, timeout);
    }
    exportProofTxs() {
        return this.getProofOutputs().map(proofOutputToProofTx);
    }
    async send() {
        const newAccountPublicKey = await this.core.derivePublicKey(this.newAccountPrivateKey);
        if (!(await this.core.userExists(newAccountPublicKey))) {
            throw new Error('Add the new account to the sdk first.');
        }
        const proofs = this.getProofOutputs();
        this.txIds = await this.core.sendProofs(proofs);
        return this.txIds[0];
    }
    getTxIds() {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        return this.txIds;
    }
    async awaitSettlement(timeout) {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        await Promise.all(this.txIds.map(txId => this.core.awaitSettlement(txId, timeout)));
    }
    getProofOutputs() {
        if (!this.proofOutput) {
            throw new Error('Call createProof() first.');
        }
        const proofOutputs = [this.proofOutput];
        if (this.requireDeposit) {
            proofOutputs.push(super.getProofOutput());
        }
        return proofOutputs;
    }
}
