// @aztec/sdk/dest/controllers/recover_account_controller.js
import { EthAddress } from '@aztec/barretenberg/address';
import { proofOutputToProofTx } from '../proofs/index.js';
import { createTxRefNo } from './create_tx_ref_no.js';
import { DepositHandler } from './deposit_handler.js';
export class RecoverAccountController extends DepositHandler {
    constructor(recoveryPayload, deposit, fee, depositor = EthAddress.ZERO, core, blockchain, provider) {
        const { recoveryData: { accountPublicKey }, } = recoveryPayload;
        super(deposit, fee, depositor, accountPublicKey, true, core, blockchain, provider);
        this.recoveryPayload = recoveryPayload;
        this.deposit = deposit;
        this.fee = fee;
        this.depositor = depositor;
        this.core = core;
        this.txIds = [];
        this.requireDeposit = !!this.publicInput.value;
    }
    async createProof(timeout) {
        const txRefNo = this.requireDeposit ? createTxRefNo() : 0;
        if (this.requireDeposit) {
            await super.createProof(txRefNo, timeout);
        }
        const { trustedThirdPartyPublicKey, recoveryPublicKey, recoveryData: { accountPublicKey, signature }, } = this.recoveryPayload;
        const proofInput = await this.core.createAccountProofInput(accountPublicKey, recoveryPublicKey, false, undefined, trustedThirdPartyPublicKey, undefined, undefined);
        proofInput.signature = signature;
        this.proofOutput = await this.core.createAccountProof(proofInput, txRefNo, timeout);
    }
    exportProofTxs() {
        return this.getProofOutputs().map(proofOutputToProofTx);
    }
    async send() {
        const proofs = this.getProofOutputs();
        this.txIds = await this.core.sendProofs(proofs);
        return this.txIds[0];
    }
    getTxIds() {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        return this.txIds;
    }
    async awaitSettlement(timeout) {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        await Promise.all(this.txIds.map(txId => this.core.awaitSettlement(txId, timeout)));
    }
    getProofOutputs() {
        if (!this.proofOutput) {
            throw new Error('Call createProof() first.');
        }
        const proofOutputs = [this.proofOutput];
        if (this.requireDeposit) {
            proofOutputs.push(super.getProofOutput());
        }
        return proofOutputs;
    }
}
