// @aztec/sdk/dest/controllers/register_controller.js
import { EthAddress } from '@aztec/barretenberg/address';
import { proofOutputToProofTx } from '../proofs/index.js';
import { SchnorrSigner } from '../signer/index.js';
import { createTxRefNo } from './create_tx_ref_no.js';
import { DepositHandler } from './deposit_handler.js';
export class RegisterController extends DepositHandler {
    constructor(userId, alias, accountPrivateKey, spendingPublicKey, recoveryPublicKey, deposit, fee, depositor = EthAddress.ZERO, core, blockchain, provider) {
        super(deposit, fee, depositor, userId, true, core, blockchain, provider);
        this.userId = userId;
        this.alias = alias;
        this.accountPrivateKey = accountPrivateKey;
        this.spendingPublicKey = spendingPublicKey;
        this.recoveryPublicKey = recoveryPublicKey;
        this.deposit = deposit;
        this.fee = fee;
        this.depositor = depositor;
        this.core = core;
        this.txIds = [];
        this.requireDeposit = !!this.publicInput.value;
    }
    async createProof(timeout) {
        const accountPublicKey = await this.core.derivePublicKey(this.accountPrivateKey);
        if (!accountPublicKey.equals(this.userId)) {
            throw new Error('`accountPrivateKey` does not belong to the user.');
        }
        const txRefNo = this.requireDeposit ? createTxRefNo() : 0;
        if (this.requireDeposit) {
            await super.createProof(txRefNo, timeout);
        }
        const proofInput = await this.core.createAccountProofInput(this.userId, accountPublicKey, false, this.alias, this.spendingPublicKey, this.recoveryPublicKey, undefined);
        const signer = new SchnorrSigner(this.core, accountPublicKey, this.accountPrivateKey);
        proofInput.signature = await signer.signMessage(proofInput.signingData);
        this.proofOutput = await this.core.createAccountProof(proofInput, txRefNo, timeout);
    }
    exportProofTxs() {
        return this.getProofOutputs().map(proofOutputToProofTx);
    }
    async send() {
        if (!(await this.core.userExists(this.userId))) {
            throw new Error('Add the user to the sdk first.');
        }
        const proofs = this.getProofOutputs();
        this.txIds = await this.core.sendProofs(proofs);
        return this.txIds[0];
    }
    getTxIds() {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        return this.txIds;
    }
    async awaitSettlement(timeout) {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutput ? 'createProof()' : 'send()'} first.`);
        }
        await Promise.all(this.txIds.map(txId => this.core.awaitSettlement(txId, timeout)));
    }
    getProofOutputs() {
        if (!this.proofOutput) {
            throw new Error('Call createProof() first.');
        }
        const proofOutputs = [this.proofOutput];
        if (this.requireDeposit) {
            proofOutputs.push(super.getProofOutput());
        }
        return proofOutputs;
    }
}
