// @aztec/sdk/dest/controllers/withdraw_controller.js
import { proofOutputToProofTx } from '../proofs/index.js';
import { createTxRefNo } from './create_tx_ref_no.js';
export class WithdrawController {
    constructor(userId, userSigner, assetValue, fee, recipient, core) {
        this.userId = userId;
        this.userSigner = userSigner;
        this.assetValue = assetValue;
        this.fee = fee;
        this.recipient = recipient;
        this.core = core;
        this.proofOutputs = [];
        this.feeProofOutputs = [];
        this.txIds = [];
        if (!assetValue.value) {
            throw new Error('Value must be greater than 0.');
        }
        this.requireFeePayingTx = !!fee.value && fee.assetId !== assetValue.assetId;
    }
    async createProof(timeout) {
        const { assetId, value } = this.assetValue;
        const privateInput = value + (!this.requireFeePayingTx ? this.fee.value : BigInt(0));
        const spendingPublicKey = this.userSigner.getPublicKey();
        const spendingKeyRequired = !spendingPublicKey.equals(this.userId);
        const proofInputs = await this.core.createPaymentProofInputs(this.userId, assetId, BigInt(0), value, privateInput, BigInt(0), BigInt(0), this.userId, spendingKeyRequired, this.recipient, spendingPublicKey, 2);
        const txRefNo = this.requireFeePayingTx || proofInputs.length > 1 ? createTxRefNo() : 0;
        if (this.requireFeePayingTx) {
            const feeProofInputs = await this.core.createPaymentProofInputs(this.userId, this.fee.assetId, BigInt(0), BigInt(0), this.fee.value, BigInt(0), BigInt(0), this.userId, spendingKeyRequired, undefined, spendingPublicKey, 2);
            this.feeProofOutputs = [];
            for (const proofInput of feeProofInputs) {
                proofInput.signature = await this.userSigner.signMessage(proofInput.signingData);
                this.feeProofOutputs.push(await this.core.createPaymentProof(proofInput, txRefNo, timeout));
            }
        }
        {
            const proofOutputs = [];
            for (const proofInput of proofInputs) {
                proofInput.signature = await this.userSigner.signMessage(proofInput.signingData);
                proofOutputs.push(await this.core.createPaymentProof(proofInput, txRefNo, timeout));
            }
            this.proofOutputs = proofOutputs;
        }
    }
    exportProofTxs() {
        if (!this.proofOutputs.length) {
            throw new Error('Call createProof() first.');
        }
        return [...this.proofOutputs, ...this.feeProofOutputs].map(proofOutputToProofTx);
    }
    async send() {
        if (!this.proofOutputs.length) {
            throw new Error('Call createProof() first.');
        }
        this.txIds = await this.core.sendProofs([...this.proofOutputs, ...this.feeProofOutputs]);
        return this.txIds[this.proofOutputs.length - 1];
    }
    getTxIds() {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutputs.length ? 'createProof()' : 'send()'} first.`);
        }
        return this.txIds;
    }
    async awaitSettlement(timeout) {
        if (!this.txIds.length) {
            throw new Error(`Call ${!this.proofOutputs.length ? 'createProof()' : 'send()'} first.`);
        }
        await Promise.all(this.txIds.map(txId => this.core.awaitSettlement(txId, timeout)));
    }
}
