// @aztec/protocol-contracts/src/protocol_contract_data.ts

    // GENERATED FILE - DO NOT EDIT. RUN `yarn generate` or `yarn generate:data`
    import { Fr } from '@aztec/foundation/fields';
    import { AztecAddress } from '@aztec/stdlib/aztec-address';

    
    export const protocolContractNames = [
      'AuthRegistry',
'ContractInstanceDeployer',
'ContractClassRegisterer',
'MultiCallEntrypoint',
'FeeJuice',
'Router'
    ] as const;

    export type ProtocolContractName = typeof protocolContractNames[number];
  

    
    export const ProtocolContractSalt: Record<ProtocolContractName, Fr> = {
      AuthRegistry: new Fr(1),
ContractInstanceDeployer: new Fr(1),
ContractClassRegisterer: new Fr(1),
MultiCallEntrypoint: new Fr(1),
FeeJuice: new Fr(1),
Router: new Fr(1)
    };
  

    
    export const ProtocolContractAddress: Record<ProtocolContractName, AztecAddress> = {
      AuthRegistry: AztecAddress.fromBigInt(1n),
ContractInstanceDeployer: AztecAddress.fromBigInt(2n),
ContractClassRegisterer: AztecAddress.fromBigInt(3n),
MultiCallEntrypoint: AztecAddress.fromBigInt(4n),
FeeJuice: AztecAddress.fromBigInt(5n),
Router: AztecAddress.fromBigInt(6n)
    };
  

    
    export const ProtocolContractLeaves = {
      AuthRegistry: Fr.fromHexString('0x0beca5ed6586f8b687c905663b17d315df05123fb7274ed26350db393d6f630a'),
ContractInstanceDeployer: Fr.fromHexString('0x180a49e656fe88966e3951bda3b00f9dec72d6059e4776907f5a1a132d802d0d'),
ContractClassRegisterer: Fr.fromHexString('0x04f20412b650b1f452c33f9a66b734bbbf715d0eebe1ae1d206e476b67f258c8'),
MultiCallEntrypoint: Fr.fromHexString('0x0d337c1f03d1fbf286b139fb3982dcf8d6ef2b0b4bf6cc618e48d8117a9c8e28'),
FeeJuice: Fr.fromHexString('0x278942aa357975cbad0ea797c2818676f6c8ffa2825f229b5bc0e5e36f7e3968'),
Router: Fr.fromHexString('0x29f39a7560a7cd71be918e823ddf7241c5d83a3d382d0cadda1a8a4b1e9edde0')
    };
  

    
    export const protocolContractTreeRoot = Fr.fromHexString('0x12edf3adf4adf6fc2f245d482883923bf92dce7a1b0c65b2b7b68a77fca9b2e6');
  

    
  export const REGISTERER_CONTRACT_CLASS_REGISTERED_TAG = Fr.fromHexString('0x20cd27645f65c15a38720ca984aef936592f3938a705db706d068ad13fc58ae9');
  export const REGISTERER_PRIVATE_FUNCTION_BROADCASTED_TAG = Fr.fromHexString('0x0abf338154e355442d9916d630da0491036d5a7b23c42b477c1395ba3bf9b42d');
  export const REGISTERER_UNCONSTRAINED_FUNCTION_BROADCASTED_TAG = Fr.fromHexString('0x27774539896271d6dbe0c1de45ddc44709d1e8d9a9fb686ca3c105d87c417bb8');
  export const DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_TAG = Fr.fromHexString('0x2ec28b91a5f838506d6042915005ff55cf7a0a5f889a83b11faed33a31b486f2');
   export const DEPLOYER_CONTRACT_INSTANCE_UPDATED_TAG = new Fr(1534834688047131268740281708431107902615560100979874281215533519862n);
