// @aztec/sdk/dest/auth_algorithms/key_pair_auth_algorithms.js
import { createJoinSplitProofSigningData } from '@aztec/barretenberg/client_proofs';
import { batchDecryptNotes, deriveNoteSecret, } from '@aztec/barretenberg/note_algorithms';
import { toJoinSplitTx } from '../proofs/proof_input/index.js';
export class KeyPairAuthAlgorithms {
    constructor(keyPair, grumpkin, noteAlgos, noteDecryptor, wasm) {
        this.keyPair = keyPair;
        this.grumpkin = grumpkin;
        this.noteAlgos = noteAlgos;
        this.noteDecryptor = noteDecryptor;
        this.wasm = wasm;
    }
    async computeValueNoteNullifier(commitment, gibberish = false) {
        const accountPrivateKey = await this.keyPair.getPrivateKey();
        return this.noteAlgos.valueNoteNullifier(commitment, accountPrivateKey, !gibberish);
    }
    async deriveNoteSecret(ecdhPubKey) {
        const accountPrivateKey = await this.keyPair.getPrivateKey();
        return deriveNoteSecret(ecdhPubKey, accountPrivateKey, this.grumpkin);
    }
    async decryptViewingKeys(viewingKeysBuf) {
        const accountPrivateKey = await this.keyPair.getPrivateKey();
        return await batchDecryptNotes(viewingKeysBuf, accountPrivateKey, this.noteDecryptor, this.grumpkin);
    }
    async createJoinSplitProofSigningData(tx) {
        const accountPrivateKey = await this.keyPair.getPrivateKey();
        return await createJoinSplitProofSigningData(toJoinSplitTx(tx, accountPrivateKey), this.wasm);
    }
}
