// @aztec/sdk/dest/database/database.js
export class SpendingKey {
    constructor(userId, key, // only contains x coordinate of a grumpkin address.
    treeIndex, hashPath) {
        this.userId = userId;
        this.key = key;
        this.treeIndex = treeIndex;
        this.hashPath = hashPath;
        if (key.length !== 32) {
            throw new Error('Invalid key buffer.');
        }
    }
}
export class Alias {
    constructor(accountPublicKey, aliasHash, index, noteCommitment1, spendingPublicKeyX) {
        this.accountPublicKey = accountPublicKey;
        this.aliasHash = aliasHash;
        this.index = index;
        this.noteCommitment1 = noteCommitment1;
        this.spendingPublicKeyX = spendingPublicKeyX;
    }
}
export class BulkUserStateUpdateData {
    constructor(updateUserArgs = [], addSpendingKeyArgs = [], upsertAccountTxArgs = [], upsertPaymentTxArgs = [], upsertDefiTxArgs = [], addNoteArgs = [], nullifyNoteArgs = []) {
        this.updateUserArgs = updateUserArgs;
        this.addSpendingKeyArgs = addSpendingKeyArgs;
        this.upsertAccountTxArgs = upsertAccountTxArgs;
        this.upsertPaymentTxArgs = upsertPaymentTxArgs;
        this.upsertDefiTxArgs = upsertDefiTxArgs;
        this.addNoteArgs = addNoteArgs;
        this.nullifyNoteArgs = nullifyNoteArgs;
    }
}
