// @aztec/barretenberg/dest/world_state/world_state.js
import { MerkleTree, MemoryMerkleTree, HashPath } from '../merkle_tree/index.js';
import { WorldStateConstants } from './world_state_constants.js';
import { createDebugLogger } from '../log/index.js';
const debug = createDebugLogger('bb:world_state');
/**
 * Wraps the merkle tree that represents the "mutable" part of the data tree. That is the tree where each leaf
 * is the root of rollups data subtree. Provides ability to compute a complete hash path when given a rollups
 * "immutable" data tree.
 *
 * It is *NOT* safe to concurrently perform reads and writes, due the underlying merkle tree implementation.
 * Even though it does its updates in a batch, the walking of the tree first makes it unsafe.
 * There is no point protecting the merkle tree with a mutex here, as in any practical context, there will be
 * several external reads that need to performed atomically. Thus, the caller is responsible for mutexing reads
 * and write appropriately. Another option would be to change the API here to e.g. `getHashPaths([])`, but
 * that maybe restrictive from a client perspective.
 */
export class WorldState {
    constructor(db, pedersen) {
        this.db = db;
        this.pedersen = pedersen;
        this.subTreeDepth = 0;
    }
    async init(subTreeDepth) {
        const subTreeSize = 1 << subTreeDepth;
        this.subTreeDepth = subTreeDepth;
        const zeroNotes = Array(subTreeSize).fill(MemoryMerkleTree.ZERO_ELEMENT);
        const subTree = await MemoryMerkleTree.new(zeroNotes, this.pedersen);
        const treeSize = WorldStateConstants.DATA_TREE_DEPTH - subTreeDepth;
        const subTreeRoot = subTree.getRoot();
        debug(`initialising data tree with depth ${treeSize} and zero element of ${subTreeRoot.toString('hex')}`);
        try {
            this.tree = await MerkleTree.fromName(this.db, this.pedersen, 'data', subTreeRoot);
        }
        catch (e) {
            this.tree = await MerkleTree.new(this.db, this.pedersen, 'data', treeSize, subTreeRoot);
        }
        this.logTreeStats();
    }
    // builds a hash path at index 0 for a 'zero' tree of the given depth
    buildZeroHashPath(depth = WorldStateConstants.DATA_TREE_DEPTH) {
        let current = MemoryMerkleTree.ZERO_ELEMENT;
        const bufs = [];
        for (let i = 0; i < depth; i++) {
            bufs.push([current, current]);
            current = this.pedersen.compress(current, current);
        }
        return new HashPath(bufs);
    }
    convertNoteIndexToSubTreeIndex(noteIndex) {
        return noteIndex >> this.subTreeDepth;
    }
    async buildFullHashPath(noteIndex, immutableHashPath) {
        const noteSubTreeIndex = this.convertNoteIndexToSubTreeIndex(noteIndex);
        const mutablePath = await this.getHashPath(noteSubTreeIndex);
        const fullHashPath = new HashPath(immutableHashPath.data.concat(mutablePath.data));
        return fullHashPath;
    }
    async insertElement(index, element) {
        const subRootIndex = this.convertNoteIndexToSubTreeIndex(index);
        await this.tree.updateElement(subRootIndex, element);
        this.logTreeStats();
    }
    async insertElements(startIndex, elements) {
        const subRootIndex = this.convertNoteIndexToSubTreeIndex(startIndex);
        await this.tree.updateElements(subRootIndex, elements);
        this.logTreeStats();
    }
    logTreeStats() {
        debug(`data size: ${this.tree.getSize()}`);
        debug(`data root: ${this.tree.getRoot().toString('hex')}`);
    }
    async syncFromDb() {
        await this.tree.syncFromDb();
    }
    async getHashPath(index) {
        return await this.tree.getHashPath(index);
    }
    getRoot() {
        return this.tree.getRoot();
    }
    getSize() {
        return this.tree.getSize();
    }
}
