// @aztec/sdk/dest/proofs/proof_output/proof_output_factory.js
import { AccountProver, JoinSplitProver, UnrolledProver } from '@aztec/barretenberg/client_proofs';
import { NetCrs } from '@aztec/barretenberg/crs';
import { randomBytes } from '@aztec/barretenberg/crypto';
import { createDebugLogger } from '@aztec/barretenberg/log';
import { MemorySerialQueue } from '../../serial_queue/index.js';
import { toJoinSplitTx, } from '../proof_input/index.js';
import { AccountProofCreator } from './account_proof_creator.js';
import { DefiProofCreator } from './defi_proof_creator.js';
import { PaymentProofCreator } from './payment_proof_creator.js';
const CREATE_PROOF_TIMEOUT = 60 * 1000;
const createTxRefNo = () => randomBytes(4).readUInt32BE(0);
export class ProofOutputFactory {
    constructor(proverless, noteAlgos, pippenger, fftFactory, barretenberg, workerPool) {
        this.proverless = proverless;
        this.noteAlgos = noteAlgos;
        this.pippenger = pippenger;
        this.fftFactory = fftFactory;
        this.barretenberg = barretenberg;
        this.workerPool = workerPool;
        this.serialQueue = new MemorySerialQueue();
        this.debug = createDebugLogger('bb:provers');
    }
    async destroy() {
        this.debug('destroying...');
        // The serial queue will cancel itself. This ensures that anything currently in the queue finishes, and ensures
        // that once the await to push() returns, nothing else is on, or can be added to the queue.
        await this.serialQueue.push(() => Promise.resolve(this.serialQueue.cancel()));
        this.debug('destroyed.');
    }
    async createProofs(proofInputs, signatures, keyPair) {
        return await this.serialQueue.push(async () => {
            // TODO - init pippenger here instead of the coreSdk.
            // await this.initProvers();
            const txRefNo = proofInputs.length > 1 ? createTxRefNo() : 0;
            const proofOutputs = [];
            for (let i = 0; i < proofInputs.length; ++i) {
                const proofInput = proofInputs[i];
                const signature = signatures[i];
                if (Object.prototype.hasOwnProperty.call(proofInput, 'partialStateSecretEphPubKey')) {
                    proofOutputs.push(await this.createDefiProof(proofInput, signature, txRefNo, keyPair));
                }
                else if (Object.prototype.hasOwnProperty.call(proofInput, 'viewingKeys')) {
                    proofOutputs.push(await this.createPaymentProof(proofInput, signature, txRefNo, keyPair));
                }
                else {
                    proofOutputs.push(await this.createAccountProof(proofInput, signature, txRefNo));
                }
            }
            return proofOutputs;
        });
    }
    // TODO - change it back to private method and only use 'createProofs'.
    async createPaymentProof({ tx, viewingKeys }, signature, txRefNo, keyPair, timeout = CREATE_PROOF_TIMEOUT) {
        await this.createJoinSplitProofCreator();
        await this.computeJoinSplitProvingKey(timeout);
        const accountPrivateKey = await keyPair.getPrivateKey();
        const proofOutput = await this.paymentProofCreator.createProof(toJoinSplitTx(tx, accountPrivateKey), viewingKeys, signature, txRefNo, timeout);
        return { ...proofOutput, outputNotes: tx.outputNotes };
    }
    // TODO - change it back to private method and only use 'createProofs'.
    async createAccountProof({ tx }, signature, txRefNo, timeout = CREATE_PROOF_TIMEOUT) {
        await this.createAccountProofCreator();
        await this.computeAccountProvingKey(timeout);
        const proofOutput = await this.accountProofCreator.createProof(tx, signature, txRefNo, timeout);
        return { ...proofOutput, outputNotes: [] };
    }
    // TODO - change it back to private method and only use 'createProofs'.
    async createDefiProof({ tx, viewingKey, partialStateSecretEphPubKey }, signature, txRefNo, keyPair, timeout = CREATE_PROOF_TIMEOUT) {
        await this.createJoinSplitProofCreator();
        await this.computeJoinSplitProvingKey(timeout);
        const accountPrivateKey = await keyPair.getPrivateKey();
        const proofOutput = await this.defiProofCreator.createProof(toJoinSplitTx(tx, accountPrivateKey), viewingKey, partialStateSecretEphPubKey, signature, txRefNo, timeout);
        return { ...proofOutput, outputNotes: tx.outputNotes };
    }
    async getCrsData(circuitSize) {
        this.debug('downloading crs data...');
        const crs = new NetCrs(circuitSize);
        await crs.init();
        this.debug('done.');
        return Buffer.from(crs.getData());
    }
    async initProvers() {
        if (this.accountProofCreator) {
            return;
        }
        const maxCircuitSize = Math.max(JoinSplitProver.getCircuitSize(), AccountProver.getCircuitSize());
        const crsData = await this.getCrsData(maxCircuitSize);
        await this.pippenger.init(crsData);
        await this.createJoinSplitProofCreator();
        await this.createAccountProofCreator();
    }
    async createJoinSplitProofCreator() {
        if (this.defiProofCreator) {
            return;
        }
        const fft = await this.fftFactory.createFft(JoinSplitProver.getCircuitSize(this.proverless));
        const unrolledProver = new UnrolledProver(this.workerPool ? this.workerPool.workers[0] : this.barretenberg, this.pippenger, fft);
        this.joinSplitProver = new JoinSplitProver(unrolledProver, this.proverless);
        this.paymentProofCreator = new PaymentProofCreator(this.joinSplitProver);
        this.defiProofCreator = new DefiProofCreator(this.joinSplitProver, this.noteAlgos);
    }
    async createAccountProofCreator() {
        if (this.accountProofCreator) {
            return;
        }
        const fft = await this.fftFactory.createFft(AccountProver.getCircuitSize(this.proverless));
        const unrolledProver = new UnrolledProver(this.workerPool ? this.workerPool.workers[0] : this.barretenberg, this.pippenger, fft);
        this.accountProver = new AccountProver(unrolledProver, this.proverless);
        this.accountProofCreator = new AccountProofCreator(this.accountProver);
    }
    async computeJoinSplitProvingKey(timeout) {
        this.debug('release account proving key...');
        await this.accountProver?.releaseKey();
        this.debug('computing join-split proving key...');
        await this.joinSplitProver.computeKey(timeout);
        this.debug('done.');
    }
    async computeAccountProvingKey(timeout) {
        this.debug('release join-split proving key...');
        await this.joinSplitProver?.releaseKey();
        this.debug('computing account proving key...');
        await this.accountProver.computeKey(timeout);
        this.debug('done.');
    }
}
