// @aztec/barretenberg/dest/note_algorithms/claim_note_tx_data.js
import { toBigIntBE, toBufferBE } from '../bigint_buffer/index.js';
import { BridgeCallData } from '../bridge_call_data/index.js';
export class ClaimNoteTxData {
    constructor(value, bridgeCallData, partialStateSecret, inputNullifier) {
        this.value = value;
        this.bridgeCallData = bridgeCallData;
        this.partialStateSecret = partialStateSecret;
        this.inputNullifier = inputNullifier;
    }
    toBuffer() {
        return Buffer.concat([
            toBufferBE(this.value, 32),
            this.bridgeCallData.toBuffer(),
            this.partialStateSecret,
            this.inputNullifier,
        ]);
    }
    equals(note) {
        return this.toBuffer().equals(note.toBuffer());
    }
    static fromBuffer(buf) {
        let dataStart = 0;
        const value = toBigIntBE(buf.slice(dataStart, dataStart + 32));
        dataStart += 32;
        const bridgeCallData = BridgeCallData.fromBuffer(buf.slice(dataStart, dataStart + 32));
        dataStart += 32;
        const partialStateSecret = buf.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const inputNullifier = buf.slice(dataStart, dataStart + 32);
        return new ClaimNoteTxData(value, bridgeCallData, partialStateSecret, inputNullifier);
    }
}
ClaimNoteTxData.EMPTY = new ClaimNoteTxData(BigInt(0), BridgeCallData.ZERO, Buffer.alloc(32), Buffer.alloc(32));
ClaimNoteTxData.SIZE = ClaimNoteTxData.EMPTY.toBuffer().length;
