// @aztec/barretenberg/dest/note_algorithms/decrypted_note.js
import { GrumpkinAddress } from '../address/index.js';
import { ViewingKeyData } from '../viewing_key/index.js';
const NOTE_SECRET_SIZE = 32;
export class DecryptedNote {
    constructor(data, ephPubKey, noteSecret) {
        this.data = data;
        this.ephPubKey = ephPubKey;
        this.noteSecret = noteSecret;
        if (noteSecret.length !== NOTE_SECRET_SIZE) {
            throw new Error('Invalid note secret buffer.');
        }
    }
    static fromBuffer(buf) {
        let dataStart = 0;
        const data = ViewingKeyData.fromBuffer(buf.slice(dataStart, dataStart + ViewingKeyData.SIZE));
        dataStart += ViewingKeyData.SIZE;
        const ephPubKey = new GrumpkinAddress(buf.slice(dataStart, dataStart + GrumpkinAddress.SIZE));
        dataStart += GrumpkinAddress.SIZE;
        const noteSecret = buf.slice(dataStart, dataStart + NOTE_SECRET_SIZE);
        return new DecryptedNote(data, ephPubKey, noteSecret);
    }
    toBuffer() {
        return Buffer.concat([this.data.toBuffer(), this.ephPubKey.toBuffer(), this.noteSecret]);
    }
}
DecryptedNote.SIZE = ViewingKeyData.SIZE + GrumpkinAddress.SIZE + NOTE_SECRET_SIZE;
