// @aztec/barretenberg/dest/crypto/schnorr/signature.js
import { randomBytes } from '../random/index.js';
export class SchnorrSignature {
    constructor(buffer) {
        this.buffer = buffer;
        if (buffer.length !== SchnorrSignature.SIZE) {
            throw new Error(`Invalid signature buffer of length ${buffer.length}.`);
        }
    }
    static isSignature(signature) {
        return /^(0x)?[0-9a-f]{128}$/i.test(signature);
    }
    static fromString(signature) {
        if (!SchnorrSignature.isSignature(signature)) {
            throw new Error(`Invalid signature string: ${signature}`);
        }
        return new SchnorrSignature(Buffer.from(signature.replace(/^0x/i, ''), 'hex'));
    }
    static randomSignature() {
        return new SchnorrSignature(randomBytes(64));
    }
    s() {
        return this.buffer.slice(0, 32);
    }
    e() {
        return this.buffer.slice(32);
    }
    toBuffer() {
        return this.buffer;
    }
    toString() {
        return `0x${this.buffer.toString('hex')}`;
    }
}
SchnorrSignature.SIZE = 64;
SchnorrSignature.EMPTY = new SchnorrSignature(Buffer.alloc(64));
