// @aztec/barretenberg/dest/block_source/index.js
import { deserializeBufferFromVector, Deserializer, numToUInt32BE, serializeBigInt, serializeBufferArrayToVector, serializeBufferToVector, serializeDate, } from '../serialize/index.js';
import { TxHash } from '../blockchain/index.js';
import { DefiInteractionEvent } from './defi_interaction_event.js';
export class Block {
    constructor(txHash, mined, rollupId, rollupSize, encodedRollupProofData, offchainTxData, interactionResult, gasUsed, gasPrice, subtreeRoot) {
        this.txHash = txHash;
        this.mined = mined;
        this.rollupId = rollupId;
        this.rollupSize = rollupSize;
        this.encodedRollupProofData = encodedRollupProofData;
        this.offchainTxData = offchainTxData;
        this.interactionResult = interactionResult;
        this.gasUsed = gasUsed;
        this.gasPrice = gasPrice;
        this.subtreeRoot = subtreeRoot;
    }
    static deserialize(buf, offset = 0) {
        const des = new Deserializer(buf, offset);
        const txHash = des.exec(TxHash.deserialize);
        const mined = des.date();
        const rollupId = des.uInt32();
        const rollupSize = des.uInt32();
        const rollupProofData = des.vector();
        const offchainTxData = des.deserializeArray(deserializeBufferFromVector);
        const interactionResult = des.deserializeArray(DefiInteractionEvent.deserialize);
        const gasUsed = des.uInt32();
        const gasPrice = des.bigInt();
        const subtreeRoot = des.vector();
        return {
            elem: new Block(txHash, mined, rollupId, rollupSize, rollupProofData, offchainTxData, interactionResult, gasUsed, gasPrice, subtreeRoot.equals(Buffer.alloc(0)) ? undefined : subtreeRoot),
            adv: des.getOffset() - offset,
        };
    }
    static fromBuffer(buf) {
        return Block.deserialize(buf).elem;
    }
    toBuffer() {
        return Buffer.concat([
            this.txHash.toBuffer(),
            serializeDate(this.mined),
            numToUInt32BE(this.rollupId),
            numToUInt32BE(this.rollupSize),
            serializeBufferToVector(this.encodedRollupProofData),
            serializeBufferArrayToVector(this.offchainTxData.map(b => serializeBufferToVector(b))),
            serializeBufferArrayToVector(this.interactionResult.map(b => b.toBuffer())),
            numToUInt32BE(this.gasUsed),
            serializeBigInt(this.gasPrice),
            serializeBufferToVector(this.subtreeRoot ?? Buffer.alloc(0)),
        ]);
    }
}
export * from './server_block_source.js';
export * from './defi_interaction_event.js';
export * from './decoded_block.js';
