// @aztec/barretenberg/dest/block_source/server_block_source.js
import { Block } from './index.js';
import { EventEmitter } from 'events';
import { fetch } from '../iso_fetch/index.js';
import { Deserializer } from '../serialize/index.js';
// import { createDebugLogger } from '@aztec/barretenberg/log';
// const debug = createDebugLogger('bb:server_block_source');
export class ServerBlockSource extends EventEmitter {
    constructor(baseUrl, pollInterval = 10000, version = '') {
        super();
        this.pollInterval = pollInterval;
        this.version = version;
        this.running = false;
        this.runningPromise = Promise.resolve();
        this.interruptPromise = Promise.resolve();
        this.interruptResolve = () => { };
        this.baseUrl = baseUrl.toString().replace(/\/$/, '');
    }
    async getLatestRollupId() {
        const url = new URL(`${this.baseUrl}/latest-rollup-id`);
        const init = this.version ? { headers: { version: this.version } } : {};
        const response = await this.awaitSucceed(() => fetch(url.toString(), init));
        return Number(await response.text());
    }
    async start(from = 0) {
        this.running = true;
        this.interruptPromise = new Promise(resolve => (this.interruptResolve = resolve));
        const emitBlocks = async () => {
            try {
                const blocks = await this.getBlocks(from);
                for (const block of blocks) {
                    this.emit('block', block);
                    from = block.rollupId + 1;
                }
            }
            catch (err) {
                // debug(err);
            }
        };
        await emitBlocks();
        const poll = async () => {
            while (this.running) {
                await emitBlocks();
                await this.sleepOrInterrupted(this.pollInterval);
            }
        };
        this.runningPromise = poll();
    }
    stop() {
        this.running = false;
        this.interruptResolve();
        return this.runningPromise;
    }
    async awaitSucceed(fn) {
        while (true) {
            try {
                const response = await fn();
                if (response.status == 409) {
                    const body = await response.json();
                    this.emit('versionMismatch', body.error);
                    throw new Error(body.error);
                }
                if (response.status !== 200) {
                    throw new Error(`Bad status code: ${response.status}`);
                }
                return response;
            }
            catch (err) {
                console.log(err.message);
                await new Promise(resolve => setTimeout(resolve, 10000));
            }
        }
    }
    async getBlocks(from, take) {
        const url = new URL(`${this.baseUrl}/get-blocks`);
        url.searchParams.append('from', from.toString());
        if (take !== undefined) {
            url.searchParams.append('take', take.toString());
        }
        const options = {
            cache: 'force-cache',
            headers: this.version ? { version: this.version } : {},
        };
        const response = await this.awaitSucceed(() => fetch(url.toString(), options));
        const result = Buffer.from(await response.arrayBuffer());
        const des = new Deserializer(result);
        return des.deserializeArray(Block.deserialize);
    }
    async sleepOrInterrupted(ms) {
        let timeout;
        const promise = new Promise(resolve => (timeout = setTimeout(resolve, ms)));
        await Promise.race([promise, this.interruptPromise]);
        clearTimeout(timeout);
    }
}
