// @aztec/barretenberg/dest/tx_id/index.js
import { randomBytes } from '../crypto/index.js';
export class TxId {
    constructor(buffer) {
        this.buffer = buffer;
        if (buffer.length !== 32) {
            throw new Error('Invalid hash buffer.');
        }
    }
    static deserialize(buffer, offset) {
        return { elem: new TxId(buffer.slice(offset, offset + 32)), adv: 32 };
    }
    static fromString(hash) {
        return new TxId(Buffer.from(hash.replace(/^0x/i, ''), 'hex'));
    }
    static random() {
        return new TxId(randomBytes(32));
    }
    equals(rhs) {
        return this.toBuffer().equals(rhs.toBuffer());
    }
    toBuffer() {
        return this.buffer;
    }
    toString() {
        return `0x${this.toBuffer().toString('hex')}`;
    }
    toDepositSigningData() {
        const digest = this.toString();
        return Buffer.concat([
            Buffer.from('Signing this message will allow your pending funds to be spent in Aztec transaction:\n\n'),
            Buffer.from(digest),
            Buffer.from('\n\nIMPORTANT: Only sign the message if you trust the client'),
        ]);
    }
}
