// @aztec/blockchain/dest/contracts/asset/eth_asset.js
/* eslint-disable @typescript-eslint/no-unused-vars */
import { EthAddress } from '@aztec/barretenberg/address';
import { TxHash } from '@aztec/barretenberg/blockchain';
import { Web3Provider } from '@ethersproject/providers';
import { fromBaseUnits, toBaseUnits } from '../../units.js';
const fixEthersStackTrace = (err) => {
    err.stack += new Error().stack;
    throw err;
};
export class EthAsset {
    constructor(provider, minConfirmations = 1) {
        this.minConfirmations = minConfirmations;
        this.provider = new Web3Provider(provider);
    }
    getStaticInfo() {
        return {
            address: EthAddress.ZERO,
            name: 'Eth',
            symbol: 'ETH',
            decimals: 18,
            gasLimit: 30000,
        };
    }
    getUserNonce(account) {
        return Promise.resolve(BigInt(0));
    }
    async balanceOf(account) {
        const balance = await this.provider.getBalance(account.toString());
        return BigInt(balance.toString());
    }
    allowance(owner, receiver) {
        throw new Error('Allowance unsupported for ETH.');
    }
    approve(value, owner, receiver) {
        throw new Error('Approve unsupported for ETH.');
    }
    mint(value, account) {
        throw new Error('Mint unsupported for ETH.');
    }
    async transfer(value, from, to, options = {}) {
        const provider = options.provider ? new Web3Provider(options.provider) : this.provider;
        const signer = provider.getSigner(from.toString());
        const tx = await signer
            .sendTransaction({
            to: to.toString(),
            value: `0x${value.toString(16)}`,
            gasLimit: options.gasLimit,
            nonce: options.nonce,
        })
            .catch(fixEthersStackTrace);
        const receipt = await tx.wait(this.minConfirmations);
        return TxHash.fromString(receipt.transactionHash);
    }
    fromBaseUnits(value, precision) {
        return fromBaseUnits(value, 18, precision);
    }
    toBaseUnits(value) {
        return toBaseUnits(value, 18);
    }
}
