// @aztec/blockchain/dest/contracts/asset/token_asset.js
import { TxHash } from '@aztec/barretenberg/blockchain';
import { Web3Provider } from '@ethersproject/providers';
import { Contract } from 'ethers';
import { fromBaseUnits, toBaseUnits } from '../../units.js';
const abi = [
    'function transfer(address to, uint256 amount) public returns(bool)',
    'function decimals() public view returns (uint8)',
    'function approve(address spender, uint256 amount) public returns (bool)',
    'function allowance(address owner, address spender) public view returns (uint256)',
    'function balanceOf(address account) public view returns (uint256)',
    'function mint(address _to, uint256 _value) public returns (bool)',
    'function name() public view returns (string)',
    'function symbol() public view returns (string)',
    'function nonces(address) public view returns(uint256)',
];
const fixEthersStackTrace = (err) => {
    err.stack += new Error().stack;
    throw err;
};
export class TokenAsset {
    constructor(info, erc20, ethereumProvider, minConfirmations = 1) {
        this.info = info;
        this.erc20 = erc20;
        this.ethereumProvider = ethereumProvider;
        this.minConfirmations = minConfirmations;
    }
    static new(info, ethereumProvider, minConfirmations) {
        const erc20 = new Contract(info.address.toString(), abi, new Web3Provider(ethereumProvider));
        return new TokenAsset(info, erc20, ethereumProvider, minConfirmations);
    }
    static async fromAddress(address, ethereumProvider, gasLimit, minConfirmations) {
        const erc20 = new Contract(address.toString(), abi, new Web3Provider(ethereumProvider));
        const info = {
            address,
            name: await erc20.name(),
            symbol: await erc20.symbol(),
            decimals: +(await erc20.decimals()),
            gasLimit,
        };
        return new TokenAsset(info, erc20, ethereumProvider, minConfirmations);
    }
    get address() {
        return this.info.address;
    }
    getStaticInfo() {
        return this.info;
    }
    async getUserNonce(account) {
        return BigInt(await this.erc20.nonces(account.toString()));
    }
    async balanceOf(account) {
        const balance = await this.erc20.balanceOf(account.toString());
        return BigInt(balance);
    }
    async allowance(owner, receiver) {
        const allowance = await this.erc20.allowance(owner.toString(), receiver.toString());
        return BigInt(allowance);
    }
    async approve(value, owner, receiver, options = {}) {
        const contract = this.getContractWithSigner(owner, options);
        const res = (await contract.approve(receiver.toString(), value).catch(fixEthersStackTrace));
        const receipt = await res.wait(this.minConfirmations);
        return TxHash.fromString(receipt.transactionHash);
    }
    async mint(value, account, options = {}) {
        const contract = this.getContractWithSigner(account, options);
        const res = await contract.mint(account.toString(), value).catch(fixEthersStackTrace);
        const receipt = await res.wait(this.minConfirmations);
        return TxHash.fromString(receipt.transactionHash);
    }
    async transfer(value, from, to, options = {}) {
        const contract = this.getContractWithSigner(from, options);
        const res = await contract.transfer(to.toString(), value).catch(fixEthersStackTrace);
        const receipt = await res.wait(this.minConfirmations);
        return TxHash.fromString(receipt.transactionHash);
    }
    fromBaseUnits(value, precision) {
        return fromBaseUnits(value, this.info.decimals, precision);
    }
    toBaseUnits(value) {
        return toBaseUnits(value, this.info.decimals);
    }
    getContractWithSigner(account, options) {
        const { provider = this.ethereumProvider, signingAddress = account } = options;
        const ethSigner = new Web3Provider(provider).getSigner(signingAddress.toString());
        return new Contract(this.info.address.toString(), abi, ethSigner);
    }
}
