// @aztec/sdk/dest/user_tx/user_defi_tx.js
import { ProofId } from '@aztec/barretenberg/client_proofs';
export var UserDefiInteractionResultState;
(function (UserDefiInteractionResultState) {
    UserDefiInteractionResultState["PENDING"] = "PENDING";
    UserDefiInteractionResultState["AWAITING_FINALISATION"] = "AWAITING_FINALISATION";
    UserDefiInteractionResultState["AWAITING_SETTLEMENT"] = "AWAITING_SETTLEMENT";
    UserDefiInteractionResultState["SETTLED"] = "SETTLED";
})(UserDefiInteractionResultState = UserDefiInteractionResultState || (UserDefiInteractionResultState = {}));
export class UserDefiTx {
    constructor(txId, userId, bridgeCallData, depositValue, fee, created, settled, interactionResult) {
        this.txId = txId;
        this.userId = userId;
        this.bridgeCallData = bridgeCallData;
        this.depositValue = depositValue;
        this.fee = fee;
        this.created = created;
        this.settled = settled;
        this.interactionResult = interactionResult;
        this.proofId = ProofId.DEFI_DEPOSIT;
    }
}
