// @aztec/barretenberg/dest/blockchain/ethereum_rpc.js
import { EthAddress } from '../address/index.js';
export class EthereumRpc {
    constructor(provider) {
        this.provider = provider;
    }
    async blockNumber() {
        const result = await this.provider.request({ method: 'eth_blockNumber' });
        return Number(result);
    }
    async getChainId() {
        const result = await this.provider.request({ method: 'eth_chainId' });
        return Number(result);
    }
    async getAccounts() {
        const result = await this.provider.request({ method: 'eth_accounts' });
        return result.map(EthAddress.fromString);
    }
    async getTransactionCount(addr) {
        const result = await this.provider.request({
            method: 'eth_getTransactionCount',
            params: [addr.toString(), 'latest'],
        });
        return Number(result);
    }
    async getBalance(addr) {
        const result = await this.provider.request({
            method: 'eth_getBalance',
            params: [addr.toString(), 'latest'],
        });
        return BigInt(result);
    }
    /**
     * TODO: Return proper type with converted properties.
     */
    async getTransactionByHash(txHash) {
        const result = await this.provider.request({ method: 'eth_getTransactionByHash', params: [txHash.toString()] });
        return result;
    }
    /**
     * TODO: Return proper type with converted properties.
     * For now just baseFeePerGas.
     */
    async getBlockByNumber(numberOrTag, fullTxs = false) {
        const result = await this.provider.request({ method: 'eth_getBlockByNumber', params: [numberOrTag, fullTxs] });
        return {
            ...result,
            baseFeePerGas: BigInt(result.baseFeePerGas),
        };
    }
}
