// @aztec/blockchain/dest/tokens/index.js
export * from './token_store.js';
export * from './mainnet_addresses.js';
import { TokenStore } from './token_store.js';
import { Web3Provider } from '@ethersproject/providers';
import { Contract } from 'ethers';
import { MainnetAddresses } from './mainnet_addresses.js';
import { ERC20, WETH9 } from '../abis.js';
const getSigner = (ethereumProvider, spender) => {
    return new Web3Provider(ethereumProvider).getSigner(spender.toString());
};
export async function purchaseTokens(tokenAddress, quantityToPurchase, maximumAmountToSpend, provider, spender, recipient) {
    const tokenStore = await TokenStore.create(provider);
    const token = {
        amount: quantityToPurchase,
        erc20Address: tokenAddress,
    };
    try {
        return await tokenStore.purchase(spender, recipient ? recipient : spender, token, maximumAmountToSpend);
    }
    catch (e) {
        console.log(e);
    }
}
export async function getTokenBalance(tokenAddress, owner, ethereumProvider) {
    const tokenContract = new Contract(tokenAddress.toString(), ERC20.abi, new Web3Provider(ethereumProvider));
    const currentBalance = await tokenContract.balanceOf(owner.toString());
    return currentBalance.toBigInt();
}
export async function getTokenAllowance(tokenAddress, owner, spender, ethereumProvider) {
    const tokenContract = new Contract(tokenAddress.toString(), ERC20.abi, new Web3Provider(ethereumProvider));
    const currentBalance = await tokenContract.allowance(owner.toString(), spender.toString());
    return currentBalance.toBigInt();
}
export async function approveToken(tokenAddress, owner, spender, ethereumProvider, amount) {
    const signer = getSigner(ethereumProvider, owner);
    const tokenContract = new Contract(tokenAddress.toString(), ERC20.abi, signer);
    const approved = await tokenContract.approve(spender.toString(), amount);
    await approved.wait();
}
export async function transferToken(tokenAddress, spender, recipient, ethereumProvider, amount) {
    const signer = getSigner(ethereumProvider, spender);
    const tokenContract = new Contract(tokenAddress.toString(), ERC20.abi, signer);
    const approved = await tokenContract.transfer(recipient.toString(), amount);
    await approved.wait();
}
export async function approveWeth(owner, spender, amount, ethereumProvider) {
    const signer = getSigner(ethereumProvider, owner);
    const wethContract = new Contract(MainnetAddresses.Tokens['WETH'], WETH9.abi, signer);
    const approveTx = await wethContract.approve(spender.toString(), amount);
    await approveTx.wait();
}
export async function getWethBalance(owner, ethereumProvider) {
    const wethContract = new Contract(MainnetAddresses.Tokens['WETH'], WETH9.abi, new Web3Provider(ethereumProvider));
    const currentBalance = await wethContract.balanceOf(owner.toString());
    return currentBalance.toBigInt();
}
export async function depositToWeth(spender, amount, ethereumProvider) {
    const signer = getSigner(ethereumProvider, spender);
    const wethContract = new Contract(MainnetAddresses.Tokens['WETH'], WETH9.abi, signer);
    const balance = (await wethContract.balanceOf(spender.toString())).toBigInt();
    if (balance < amount) {
        const amountToAdd = amount - balance;
        const depositTx = await wethContract.deposit({ value: amountToAdd });
        await depositTx.wait();
    }
}
