// @aztec/barretenberg/dest/crypto/pedersen/single_pedersen.js
import { deserializeArrayFromVector, deserializeField, serializeBufferArrayToVector } from '../../serialize/index.js';
import { BarretenbergWasm } from '../../wasm/index.js';
import { Transfer } from '../../transport/index.js';
/**
 * Single threaded implementation of pedersen.
 */
export class SinglePedersen {
    /**
     * Long running functions can execute on a worker. If none is provided, call the wasm on the calling thread.
     *
     * @param wasm Synchronous functions will use use this wasm directly on the calling thread.
     * @param worker Asynchronous functions execute on this worker, preventing blocking the calling thread.
     */
    constructor(wasm, worker = wasm) {
        this.wasm = wasm;
        this.worker = worker;
    }
    async init() {
        this.wasm.call('pedersen__init');
        await this.worker.call('pedersen__init');
    }
    compress(lhs, rhs) {
        this.wasm.transferToHeap(lhs, 0);
        this.wasm.transferToHeap(rhs, 32);
        this.wasm.call('pedersen__compress_fields', 0, 32, 64);
        return Buffer.from(this.wasm.sliceMemory(64, 96));
    }
    compressInputs(inputs) {
        const inputVectors = serializeBufferArrayToVector(inputs);
        this.wasm.transferToHeap(inputVectors, 0);
        this.wasm.call('pedersen__compress', 0, 0);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    compressWithHashIndex(inputs, hashIndex) {
        const inputVectors = serializeBufferArrayToVector(inputs);
        this.wasm.transferToHeap(inputVectors, 0);
        this.wasm.call('pedersen__compress_with_hash_index', 0, 0, hashIndex);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    hashToField(data) {
        const mem = this.wasm.call('bbmalloc', data.length);
        this.wasm.transferToHeap(data, mem);
        this.wasm.call('pedersen__buffer_to_field', mem, data.length, 0);
        this.wasm.call('bbfree', mem);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    async hashToTree(values) {
        const data = serializeBufferArrayToVector(values);
        const inputPtr = await this.worker.call('bbmalloc', data.length);
        await this.worker.transferToHeap(this.worker instanceof BarretenbergWasm ? data : Transfer(data, [data.buffer]), inputPtr);
        const resultPtr = await this.worker.call('pedersen__hash_to_tree', inputPtr);
        const resultNumFields = Buffer.from(await this.worker.sliceMemory(resultPtr, resultPtr + 4)).readUInt32BE(0);
        const resultData = Buffer.from(await this.worker.sliceMemory(resultPtr, resultPtr + 4 + resultNumFields * 32));
        await this.worker.call('bbfree', inputPtr);
        await this.worker.call('bbfree', resultPtr);
        return deserializeArrayFromVector(deserializeField, resultData).elem;
    }
}
