// @aztec/sdk/dest/core_tx/core_defi_tx.js
import { GrumpkinAddress } from '@aztec/barretenberg/address';
import { BridgeCallData } from '@aztec/barretenberg/bridge_call_data';
import { ProofId } from '@aztec/barretenberg/client_proofs';
import { TxId } from '@aztec/barretenberg/tx_id';
export class CoreDefiTx {
    constructor(txId, userId, bridgeCallData, depositValue, txFee, txRefNo, created, partialState, partialStateSecret, 
    // Optional because not known at the time of proof construction.
    nullifier, settled, interactionNonce, isAsync, 
    // Optional because filled in once interaction finalised.
    success, outputValueA, outputValueB, finalised, 
    // Optional becaused filled in once claimed.
    claimSettled, claimTxId) {
        this.txId = txId;
        this.userId = userId;
        this.bridgeCallData = bridgeCallData;
        this.depositValue = depositValue;
        this.txFee = txFee;
        this.txRefNo = txRefNo;
        this.created = created;
        this.partialState = partialState;
        this.partialStateSecret = partialStateSecret;
        this.nullifier = nullifier;
        this.settled = settled;
        this.interactionNonce = interactionNonce;
        this.isAsync = isAsync;
        this.success = success;
        this.outputValueA = outputValueA;
        this.outputValueB = outputValueB;
        this.finalised = finalised;
        this.claimSettled = claimSettled;
        this.claimTxId = claimTxId;
        this.proofId = ProofId.DEFI_DEPOSIT;
    }
}
export const coreDefiTxToJson = (tx) => ({
    ...tx,
    txId: tx.txId.toString(),
    userId: tx.userId.toString(),
    bridgeCallData: tx.bridgeCallData.toString(),
    depositValue: tx.depositValue.toString(),
    txFee: tx.txFee.toString(),
    partialState: tx.partialState.toString('hex'),
    partialStateSecret: tx.partialStateSecret.toString('hex'),
    nullifier: tx.nullifier?.toString('hex'),
    outputValueA: tx.outputValueA?.toString(),
    outputValueB: tx.outputValueB?.toString(),
    claimTxId: tx.claimTxId?.toString(),
});
export const coreDefiTxFromJson = (json) => new CoreDefiTx(TxId.fromString(json.txId), GrumpkinAddress.fromString(json.userId), BridgeCallData.fromString(json.bridgeCallData), BigInt(json.depositValue), BigInt(json.txFee), json.txRefNo, json.created, Buffer.from(json.partialState, 'hex'), Buffer.from(json.partialStateSecret, 'hex'), json.nullifier ? Buffer.from(json.nullifier, 'hex') : undefined, json.settled, json.interactionNonce, json.isAsync, json.success, json.outputValueA ? BigInt(json.outputValueA) : undefined, json.outputValueB ? BigInt(json.outputValueB) : undefined, json.finalised, json.claimSettled, json.claimTxId ? TxId.fromString(json.claimTxId) : undefined);
