// @aztec/sdk/dest/core_tx/core_payment_tx.js
import { EthAddress, GrumpkinAddress } from '@aztec/barretenberg/address';
import { TxId } from '@aztec/barretenberg/tx_id';
/**
 * Comprises data which will be stored in the user's db.
 * Note: we must be able to restore output notes (etc.) without relying on the db
 * (since local browser data might be cleared, or the user might login from other devices),
 * so crucial data which enables such restoration must not be solely stored here;
 * it must also be contained in either the viewingKey or the offchainTxData.
 */
export class CorePaymentTx {
    constructor(txId, userId, proofId, assetId, publicValue, publicOwner, privateInput, recipientPrivateOutput, senderPrivateOutput, isRecipient, isSender, txRefNo, created, settled) {
        this.txId = txId;
        this.userId = userId;
        this.proofId = proofId;
        this.assetId = assetId;
        this.publicValue = publicValue;
        this.publicOwner = publicOwner;
        this.privateInput = privateInput;
        this.recipientPrivateOutput = recipientPrivateOutput;
        this.senderPrivateOutput = senderPrivateOutput;
        this.isRecipient = isRecipient;
        this.isSender = isSender;
        this.txRefNo = txRefNo;
        this.created = created;
        this.settled = settled;
    }
}
export const corePaymentTxToJson = (tx) => ({
    ...tx,
    txId: tx.txId.toString(),
    userId: tx.userId.toString(),
    publicValue: tx.publicValue.toString(),
    publicOwner: tx.publicOwner ? tx.publicOwner.toString() : undefined,
    privateInput: tx.privateInput.toString(),
    recipientPrivateOutput: tx.recipientPrivateOutput.toString(),
    senderPrivateOutput: tx.senderPrivateOutput.toString(),
});
export const corePaymentTxFromJson = (json) => new CorePaymentTx(TxId.fromString(json.txId), GrumpkinAddress.fromString(json.userId), json.proofId, json.assetId, BigInt(json.publicValue), json.publicOwner ? EthAddress.fromString(json.publicOwner) : undefined, BigInt(json.privateInput), BigInt(json.recipientPrivateOutput), BigInt(json.senderPrivateOutput), json.isRecipient, json.isSender, json.txRefNo, json.created, json.settled);
export const createCorePaymentTxForRecipient = ({ txId, userId, proofId, assetId, publicValue, publicOwner, privateInput, recipientPrivateOutput, senderPrivateOutput, txRefNo, created, settled, }, recipient) => new CorePaymentTx(txId, recipient, proofId, assetId, publicValue, publicOwner, privateInput, recipientPrivateOutput, senderPrivateOutput, true, recipient.equals(userId), txRefNo, created, settled);
