// @aztec/sdk/dest/core_tx/core_account_tx.js
import { AliasHash } from '@aztec/barretenberg/account_id';
import { GrumpkinAddress } from '@aztec/barretenberg/address';
import { ProofId } from '@aztec/barretenberg/client_proofs';
import { TxId } from '@aztec/barretenberg/tx_id';
export class CoreAccountTx {
    constructor(txId, userId, aliasHash, newSpendingPublicKey1, newSpendingPublicKey2, migrated, txRefNo, created, settled) {
        this.txId = txId;
        this.userId = userId;
        this.aliasHash = aliasHash;
        this.newSpendingPublicKey1 = newSpendingPublicKey1;
        this.newSpendingPublicKey2 = newSpendingPublicKey2;
        this.migrated = migrated;
        this.txRefNo = txRefNo;
        this.created = created;
        this.settled = settled;
        this.proofId = ProofId.ACCOUNT;
    }
}
export const coreAccountTxToJson = (tx) => ({
    ...tx,
    txId: tx.txId.toString(),
    userId: tx.userId.toString(),
    aliasHash: tx.aliasHash.toString(),
    newSpendingPublicKey1: tx.newSpendingPublicKey1 ? tx.newSpendingPublicKey1.toString('hex') : undefined,
    newSpendingPublicKey2: tx.newSpendingPublicKey2 ? tx.newSpendingPublicKey2.toString('hex') : undefined,
});
export const coreAccountTxFromJson = (json) => new CoreAccountTx(TxId.fromString(json.txId), GrumpkinAddress.fromString(json.userId), AliasHash.fromString(json.aliasHash), json.newSpendingPublicKey1 ? Buffer.from(json.newSpendingPublicKey1, 'hex') : undefined, json.newSpendingPublicKey2 ? Buffer.from(json.newSpendingPublicKey2, 'hex') : undefined, json.migrated, json.txRefNo, json.created, json.settled);
