// @aztec/sdk/dest/note/note.js
import { TreeNote } from '@aztec/barretenberg/note_algorithms';
export class Note {
    constructor(treeNote, commitment, nullifier, allowChain, nullified, index, hashPath) {
        this.treeNote = treeNote;
        this.commitment = commitment;
        this.nullifier = nullifier;
        this.allowChain = allowChain;
        this.nullified = nullified;
        this.index = index;
        this.hashPath = hashPath;
    }
    get assetId() {
        return this.treeNote.assetId;
    }
    get value() {
        return this.treeNote.value;
    }
    get owner() {
        return this.treeNote.ownerPubKey;
    }
    get ownerAccountRequired() {
        return this.treeNote.accountRequired;
    }
    get pending() {
        return this.index === undefined;
    }
}
export const noteToJson = ({ treeNote, commitment, nullifier, allowChain, nullified, index, hashPath, }) => ({
    treeNote: new Uint8Array(treeNote.toBuffer()),
    commitment: commitment.toString('hex'),
    nullifier: nullifier.toString('hex'),
    allowChain,
    nullified,
    index,
    hashPath: hashPath?.toString('hex'),
});
export const noteFromJson = ({ treeNote, commitment, nullifier, allowChain, nullified, index, hashPath }) => new Note(TreeNote.fromBuffer(Buffer.from(treeNote)), Buffer.from(commitment, 'hex'), Buffer.from(nullifier, 'hex'), allowChain, nullified, index, hashPath ? Buffer.from(hashPath, 'hex') : undefined);
