// @aztec/blockchain/dest/provider/wallet_provider.js
import { Wallet } from 'ethers';
import { EthAddress } from '@aztec/barretenberg/address';
import { EthersAdapter } from './ethers_adapter.js';
import { JsonRpcProvider, Web3Provider } from '@ethersproject/providers';
import { readFileSync } from 'fs';
/**
 * Given an EIP1193 provider, wraps it, and provides the ability to add local accounts.
 */
export class WalletProvider {
    constructor(provider) {
        this.provider = provider;
        this.accounts = [];
    }
    static fromHost(ethereumHost) {
        const ethersProvider = new JsonRpcProvider(ethereumHost);
        return new WalletProvider(new EthersAdapter(ethersProvider));
    }
    addAccount(privateKey) {
        return this.addEthersWallet(new Wallet(privateKey, new Web3Provider(this.provider)));
    }
    addAccountsFromMnemonic(mnemonic, num, bip32Account = 0) {
        for (let i = 0; i < num; ++i) {
            this.addAccountFromMnemonicAndPath(mnemonic, `m/44'/60'/${bip32Account}'/0/${i}`);
        }
    }
    addAccountFromMnemonicAndPath(mnemonic, path) {
        return this.addEthersWallet(Wallet.fromMnemonic(mnemonic, path).connect(new Web3Provider(this.provider)));
    }
    addAccountFromKeystore(file, password = '') {
        const json = readFileSync(file, { encoding: 'ascii' });
        this.addEthersWallet(Wallet.fromEncryptedJsonSync(json, password));
    }
    addEthersWallet(wallet) {
        this.accounts.push(wallet);
        return EthAddress.fromString(wallet.address);
    }
    getAccounts() {
        return this.accounts.map(a => EthAddress.fromString(a.address));
    }
    getAccount(account) {
        return EthAddress.fromString(this.accounts[account].address);
    }
    getPrivateKey(account) {
        return Buffer.from(this.accounts[account].privateKey.slice(2), 'hex');
    }
    getPrivateKeyForAddress(account) {
        const wallet = this.accounts.find(w => account.equals(EthAddress.fromString(w.address)));
        return wallet ? Buffer.from(wallet.privateKey.slice(2), 'hex') : undefined;
    }
    async request(args) {
        switch (args.method) {
            case 'eth_accounts':
                return this.accounts.length ? this.accounts.map(a => a.address) : await this.provider.request(args);
            case 'eth_sign':
                return await this.sign(args);
            case 'personal_sign':
                return await this.personalSign(args);
            case 'eth_signTypedData_v4':
                return this.signTypedData(args);
            case 'eth_signTransaction':
                return this.signTransaction(args);
            case 'eth_sendTransaction':
                return this.sendTransaction(args);
            default: {
                return await this.provider.request(args);
            }
        }
    }
    async personalSign(args) {
        const [message, from] = args.params;
        const account = this.accounts.find(a => a.address.toLowerCase() === from.toLowerCase());
        if (account) {
            return await account.signMessage(Buffer.from(message.slice(2), 'hex'));
        }
        return await this.provider.request(args);
    }
    async sign(args) {
        const [from, message] = args.params;
        const account = this.accounts.find(a => a.address.toLowerCase() === from.toLowerCase());
        if (account) {
            return await account.signMessage(Buffer.from(message.slice(2), 'hex'));
        }
        return await this.provider.request(args);
    }
    async signTypedData(args) {
        const [from, data] = args.params;
        const { types, domain, message } = typeof data === 'string' ? JSON.parse(data) : data;
        const account = this.accounts.find(a => a.address.toLowerCase() === from.toLowerCase());
        if (account) {
            delete types.EIP712Domain;
            return await account._signTypedData(domain, types, message);
        }
        return this.provider.request(args);
    }
    /**
     * Given a tx in Eth Json Rpc format, convert to ethers format and give to account to sign.
     * Populate any missing fields.
     */
    async signTxLocally(tx, account) {
        const { gas = null, value = 0, from, to, data, nonce = null } = tx;
        const txReq = {
            from,
            to,
            data,
            gasLimit: gas,
            value,
            nonce,
        };
        const toSign = await account.populateTransaction(txReq);
        return await account.signTransaction(toSign);
    }
    async signTransaction(args) {
        const tx = args.params[0];
        const account = this.accounts.find(a => a.address.toLowerCase() === tx.from.toLowerCase());
        if (account) {
            return this.signTxLocally(tx, account);
        }
        return await this.provider.request(args);
    }
    async sendTransaction(args) {
        const tx = args.params[0];
        const account = this.accounts.find(a => a.address.toLowerCase() === tx.from.toLowerCase());
        if (account) {
            const result = await this.signTxLocally(tx, account);
            return this.provider.request({ method: 'eth_sendRawTransaction', params: [result] });
        }
        return this.provider.request(args);
    }
    on(notification, listener) {
        return this.provider.on(notification, listener);
    }
    removeListener(notification, listener) {
        return this.provider.removeListener(notification, listener);
    }
}
