// @aztec/blockchain/dest/contracts/decode_error.js
import { utils } from 'ethers';
function extractFragmentsBySelector(contract, filter) {
    const errorMappings = {};
    const fragments = filter ? contract.interface.fragments.filter(f => f.type === filter) : contract.interface.fragments;
    for (const frag of fragments) {
        const sig = `${frag.name}(${frag.inputs.map(f => f.type).join(',')})`;
        const fullHash = utils.keccak256(utils.toUtf8Bytes(sig));
        const selector = fullHash.slice(2, 10);
        errorMappings[selector] = {
            name: frag.name,
            type: frag.type,
            fullHash,
            inputs: frag.inputs.map(f => f.type),
        };
    }
    return errorMappings;
}
export function decodeSelector(contract, selector) {
    const mappings = extractFragmentsBySelector(contract);
    return mappings[selector];
}
export function retrieveContractSelectors(contract, type) {
    return extractFragmentsBySelector(contract, type);
}
export function decodeErrorFromContract(contract, data) {
    const errorMappings = extractFragmentsBySelector(contract, 'error');
    const fullData = data.slice(2);
    const sigHash = fullData.slice(0, 8);
    // the rest of the data is any arguments given to the revert
    const args = fullData.slice(8);
    // look to see if we have the signature hash
    if (!errorMappings[sigHash]) {
        return;
    }
    const errorMapping = errorMappings[sigHash];
    // now try and decode the params based on their input type
    let result = [];
    if (errorMapping && errorMapping.inputs.length) {
        const abiDecodeString = errorMapping.inputs;
        result = contract.interface._abiCoder.decode(abiDecodeString, `0x${args}`).map(x => x.toString());
    }
    const errorValue = {
        name: errorMapping.name,
        params: result,
    };
    return errorValue;
}
export async function decodeErrorFromContractByTxHash(contract, txHash, provider) {
    const { to, from, gas, maxFeePerGas, maxPriorityFeePerGas, input, value, nonce, blockNumber } = await provider.request({
        method: 'eth_getTransactionByHash',
        params: [txHash.toString()],
    });
    const req = {
        to,
        from,
        gas,
        maxFeePerGas,
        maxPriorityFeePerGas,
        data: input,
        value,
        nonce,
    };
    const rep = await provider.request({ method: 'eth_call', params: [req, blockNumber] }).catch(err => err);
    if (!rep.data) {
        return;
    }
    return decodeErrorFromContract(contract, rep.data);
}
