// @aztec/sdk/dest/note_picker/index.js
const noteSum = (notes) => notes.reduce((sum, { value }) => sum + value, BigInt(0));
export class NotePicker {
    constructor(notes = []) {
        // Filter non spendable
        const validNotes = notes.filter(note => !note.pending || note.allowChain);
        this.sortedNotes = {
            pendingAreLower: this.sortNotes(validNotes, true),
            pendingAreHigher: this.sortNotes(validNotes, false),
        };
    }
    pick(value, { excludedNullifiers, excludePendingNotes, spendingKeyRequired } = {}) {
        const filteredNotes = this.filterByOptions(this.getSortedNotes(true), {
            excludedNullifiers,
            excludePendingNotes,
            spendingKeyRequired,
        });
        let minimumNumberOfNotes = this.findMinimumNotes(value, filteredNotes);
        if (!minimumNumberOfNotes) {
            // Not enough notes
            return [];
        }
        // We can pick 2 notes with the same cost
        if (minimumNumberOfNotes === 1) {
            minimumNumberOfNotes = 2;
        }
        let pickedNotes = this.getMaxSpendableNotes(excludedNullifiers, excludePendingNotes, spendingKeyRequired, minimumNumberOfNotes).reverse();
        pickedNotes = this.reducePickedNotesValues(pickedNotes, value, {
            excludedNullifiers,
            excludePendingNotes,
            spendingKeyRequired,
        });
        const exactNoteMatch = pickedNotes.find(note => note.value === value);
        if (exactNoteMatch) {
            return [exactNoteMatch];
        }
        else {
            return pickedNotes;
        }
    }
    reducePickedNotesValues(selectionToImprove, targetValue, options) {
        const pickedNotes = selectionToImprove.slice();
        const availableNotes = this.filterByOptions(this.getSortedNotes(false), options);
        let excessValue = 0n;
        let usedNullifiers = [];
        let unusedNotes = [];
        let hasPending = false;
        for (let index = 0; index < pickedNotes.length; index++) {
            const noteToImprove = pickedNotes[index];
            excessValue = noteSum(pickedNotes) - targetValue;
            usedNullifiers = pickedNotes.map(note => note.nullifier);
            unusedNotes = availableNotes.filter(note => !usedNullifiers.includes(note.nullifier));
            hasPending = pickedNotes.some(note => note.pending);
            const minimumValue = noteToImprove.value - excessValue;
            const betterNoteIndex = unusedNotes.findIndex(potentallyBetterNote => potentallyBetterNote.value >= minimumValue &&
                potentallyBetterNote.value < noteToImprove.value &&
                (!potentallyBetterNote.pending || !hasPending || noteToImprove.pending));
            if (betterNoteIndex === -1) {
                // No better note found, cannot improve further the set
                return pickedNotes;
            }
            const betterNote = unusedNotes[betterNoteIndex];
            pickedNotes[index] = betterNote;
        }
        return pickedNotes;
    }
    findMinimumNotes(value, notes) {
        let minimumNotes = 0;
        let accumulatedValue = 0n;
        let hasPendingNote = false;
        for (let index = notes.length - 1; index >= 0; index--) {
            const note = notes[index];
            if (!note.pending || !hasPendingNote) {
                minimumNotes++;
                accumulatedValue += note.value;
                hasPendingNote = hasPendingNote || note.pending;
                if (accumulatedValue >= value) {
                    return minimumNotes;
                }
            }
        }
        return 0;
    }
    pickOne(value, options = {}) {
        const filteredNotes = this.filterByOptions(this.getSortedNotes(false), options);
        return filteredNotes.find(note => note.value >= value);
    }
    getSum() {
        return noteSum(this.getSortedNotes().filter(note => !note.pending));
    }
    getSpendableNoteValues(options = {}) {
        return this.filterByOptions(this.getSortedNotes(true), options).map(note => note.value);
    }
    getMaxSpendableNoteValues({ excludedNullifiers, excludePendingNotes, spendingKeyRequired, numNotes, } = {}) {
        const notes = this.getMaxSpendableNotes(excludedNullifiers, excludePendingNotes, spendingKeyRequired, numNotes);
        return notes.map(n => n.value);
    }
    getMaxSpendableNotes(excludedNullifiers = [], excludePendingNotes = false, spendingKeyRequired = false, numNotes) {
        const filteredNotes = this.filterByOptions(this.getSortedNotes(true), {
            excludedNullifiers,
            excludePendingNotes,
            spendingKeyRequired,
        });
        let hasPendingNote = false;
        const notes = [];
        for (const note of filteredNotes.reverse()) {
            if (!note.pending || !hasPendingNote) {
                notes.push(note);
                hasPendingNote = hasPendingNote || note.pending;
            }
            if (notes.length === numNotes) {
                return notes;
            }
        }
        return notes;
    }
    getSortedNotes(pendingAreLower = false) {
        return pendingAreLower ? this.sortedNotes.pendingAreLower : this.sortedNotes.pendingAreHigher;
    }
    sortNotes(notes, pendingAreLower) {
        return notes.slice().sort((a, b) => {
            if (a.value < b.value) {
                return -1;
            }
            if (a.value > b.value) {
                return 1;
            }
            const pendingDiff = Number(a.pending) - Number(b.pending);
            return pendingAreLower ? -pendingDiff : pendingDiff;
        });
    }
    filterByOptions(notes, { excludedNullifiers = [], excludePendingNotes = false, spendingKeyRequired = false } = {}) {
        return notes.filter(note => !excludedNullifiers.some(n => n.equals(note.nullifier)) &&
            (!excludePendingNotes || !note.pending) &&
            (spendingKeyRequired ? note.ownerAccountRequired : !note.ownerAccountRequired));
    }
}
