// @aztec/barretenberg/dest/merkle_tree/hash_path.js
import { deserializeArrayFromVector, serializeBufferArrayToVector } from '../serialize/index.js';
export class HashPath {
    constructor(data = []) {
        this.data = data;
    }
    static ZERO(size, zeroElement, pedersen) {
        const bufs = [];
        let current = zeroElement;
        for (let i = 0; i < size; ++i) {
            bufs.push([current, current]);
            current = pedersen.compress(current, current);
        }
        return new HashPath(bufs);
    }
    toBuffer() {
        const elements = this.data.map(nodes => Buffer.concat([nodes[0], nodes[1]]));
        return serializeBufferArrayToVector(elements);
    }
    static fromBuffer(buf, offset = 0) {
        const { elem } = HashPath.deserialize(buf, offset);
        return elem;
    }
    static deserialize(buf, offset = 0) {
        const deserializePath = (buf, offset) => ({
            elem: [buf.slice(offset, offset + 32), buf.slice(offset + 32, offset + 64)],
            adv: 64,
        });
        const { elem, adv } = deserializeArrayFromVector(deserializePath, buf, offset);
        return { elem: new HashPath(elem), adv };
    }
    // For json serialization
    toString() {
        return this.toBuffer().toString('hex');
    }
    // For json deserialization
    static fromString(repr) {
        return HashPath.fromBuffer(Buffer.from(repr, 'hex'));
    }
}
