// @aztec/sdk/dest/database/node/transformer.js
import { AliasHash } from '@aztec/barretenberg/account_id';
import { EthAddress, GrumpkinAddress } from '@aztec/barretenberg/address';
import { BridgeCallData } from '@aztec/barretenberg/bridge_call_data';
import { TxId } from '@aztec/barretenberg/tx_id';
export const bigintTransformer = {
    to: (entityValue) => (entityValue !== undefined ? `${entityValue}` : ''),
    from: (dbValue) => (dbValue ? BigInt(dbValue) : undefined),
};
export const grumpkinAddressTransformer = {
    to: (entityValue) => entityValue?.toBuffer(),
    from: (dbValue) => (dbValue ? new GrumpkinAddress(dbValue) : undefined),
};
export const aliasHashTransformer = {
    to: (entityValue) => entityValue?.toBuffer(),
    from: (dbValue) => (dbValue ? new AliasHash(dbValue) : undefined),
};
export const txIdTransformer = {
    to: (entityValue) => entityValue?.toBuffer(),
    from: (dbValue) => (dbValue ? new TxId(dbValue) : undefined),
};
export const ethAddressTransformer = {
    to: (entityValue) => entityValue?.toBuffer(),
    from: (dbValue) => (dbValue ? new EthAddress(dbValue) : undefined),
};
export const bridgeCallDataTransformer = {
    to: (entityValue) => entityValue?.toBuffer(),
    from: (dbValue) => (dbValue ? BridgeCallData.fromBuffer(dbValue) : undefined),
};
